{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hi\n",
      "Done genes\n",
      "hi\n",
      "genes       COG1522  COG0859  COG0483  COG1066  COG0034  COG2217  COG0101  \\\n",
      "genomes                                                                     \n",
      "2505679030      1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "651053067       1.0      0.0      1.0      1.0      1.0      1.0      1.0   \n",
      "637000240       1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "637000208       1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "650716076       1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "637000053       1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "637000192       1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "651324080       1.0      0.0      1.0      1.0      1.0      1.0      1.0   \n",
      "2571042743      1.0      0.0      1.0      1.0      1.0      1.0      1.0   \n",
      "2574180005      1.0      0.0      1.0      1.0      1.0      1.0      1.0   \n",
      "\n",
      "genes       COG0548  COG0642  COG1032  ...  COG5153  COG1967  COG1148  \\\n",
      "genomes                                ...                              \n",
      "2505679030      1.0      1.0      0.0  ...      0.0      0.0      0.0   \n",
      "651053067       1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "637000240       1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "637000208       1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "650716076       1.0      1.0      0.0  ...      0.0      0.0      0.0   \n",
      "637000053       1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "637000192       1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "651324080       1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "2571042743      1.0      1.0      0.0  ...      0.0      0.0      0.0   \n",
      "2574180005      1.0      1.0      0.0  ...      0.0      0.0      0.0   \n",
      "\n",
      "genes       COG3054  COG1153  COG4071  COG2214  COG5256  COG3388  Y  \n",
      "genomes                                                              \n",
      "2505679030      0.0      0.0      0.0      0.0      0.0      0.0  0  \n",
      "651053067       0.0      0.0      0.0      0.0      0.0      0.0  0  \n",
      "637000240       0.0      0.0      0.0      0.0      0.0      0.0  0  \n",
      "637000208       0.0      0.0      0.0      0.0      0.0      0.0  0  \n",
      "650716076       0.0      0.0      0.0      0.0      0.0      0.0  0  \n",
      "637000053       0.0      0.0      0.0      0.0      0.0      0.0  0  \n",
      "637000192       0.0      0.0      0.0      0.0      0.0      0.0  0  \n",
      "651324080       0.0      0.0      0.0      0.0      0.0      0.0  0  \n",
      "2571042743      0.0      0.0      0.0      0.0      0.0      0.0  0  \n",
      "2574180005      0.0      0.0      0.0      0.0      0.0      0.0  0  \n",
      "\n",
      "[10 rows x 4212 columns]\n",
      "3837\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from sklearn.feature_selection import SelectKBest\n",
    "from sklearn.feature_selection import chi2\n",
    "from sklearn.feature_selection import VarianceThreshold\n",
    "\n",
    "import os\n",
    "# Constants/filenames\n",
    "gene_index = 4\n",
    "genome_index = 1\n",
    "cogfile = 'cogs_pfams_tigrfam_ko/cogs.txt'\n",
    "\n",
    "def genomes_genes(file, genome_index, gene_index):\n",
    "    unique_genomes = []\n",
    "    unique_genes = []\n",
    "    print('hi')\n",
    "    with open(file, 'r+') as fh:\n",
    "        for line in fh.readlines()[1:]:\n",
    "            if line.split()[genome_index] not in unique_genomes:\n",
    "                unique_genomes.append(line.split()[genome_index])\n",
    "            if line.split()[gene_index] not in unique_genes:\n",
    "                unique_genes.append(line.split()[gene_index])\n",
    "    print(\"Done genes\")\n",
    "    return [unique_genomes, unique_genes]\n",
    "    \n",
    "def matrix(file, genes, genomes, gene_col, genome_col):\n",
    "    gcols = pd.Index(genes, name = \"genes\")\n",
    "    grows = pd.Index(genomes, name = \"genomes\")\n",
    "    df = pd.DataFrame(data = np.zeros(shape = (len(genomes), len(genes))), index = grows, columns = gcols)\n",
    "    fh = open(file, 'r+')\n",
    "    for line in fh.readlines()[1:]:\n",
    "        genome = line.split()[genome_index]\n",
    "        gene = line.split()[gene_index]\n",
    "        df.at[genome, gene] = 1\n",
    "    fh.close()\n",
    "    print('hi')\n",
    "    return df\n",
    "gng = genomes_genes(cogfile, genome_index, gene_index)\n",
    "allgenomes = gng[0]\n",
    "allgenes = gng[1]\n",
    "feature_matrix = matrix(cogfile, allgenes, allgenomes, gene_index, genome_index)\n",
    "feature_matrix['Y'] = [0] * len(feature_matrix)\n",
    "print(feature_matrix.head(10))\n",
    "print(len(feature_matrix))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(feature_matrix.head(12))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "genes       COG1522  COG0859  COG0483  COG1066  COG0034  COG2217  COG0101  \\\n",
      "genomes                                                                     \n",
      "2505679030      1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "651053067       1.0      0.0      1.0      1.0      1.0      1.0      1.0   \n",
      "637000240       1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "637000208       1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "650716076       1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "\n",
      "genes       COG0548  COG0642  COG1032  ...  COG5153  COG1967  COG1148  \\\n",
      "genomes                                ...                              \n",
      "2505679030      1.0      1.0      0.0  ...      0.0      0.0      0.0   \n",
      "651053067       1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "637000240       1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "637000208       1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "650716076       1.0      1.0      0.0  ...      0.0      0.0      0.0   \n",
      "\n",
      "genes       COG3054  COG1153  COG4071  COG2214  COG5256  COG3388  \\\n",
      "genomes                                                            \n",
      "2505679030      0.0      0.0      0.0      0.0      0.0      0.0   \n",
      "651053067       0.0      0.0      0.0      0.0      0.0      0.0   \n",
      "637000240       0.0      0.0      0.0      0.0      0.0      0.0   \n",
      "637000208       0.0      0.0      0.0      0.0      0.0      0.0   \n",
      "650716076       0.0      0.0      0.0      0.0      0.0      0.0   \n",
      "\n",
      "genes                                                       Y  \n",
      "genomes                                                        \n",
      "2505679030  [Bacteroidetes, Bacteroidetes, Cytophagia, Cyt...  \n",
      "651053067   [Alphaproteobacteria, Proteobacteria, Alphapro...  \n",
      "637000240   [Alphaproteobacteria, Proteobacteria, Alphapro...  \n",
      "637000208   [Burkholderiales, Proteobacteria, Betaproteoba...  \n",
      "650716076   [Pseudomonas, Proteobacteria, Gammaproteobacte...  \n",
      "\n",
      "[5 rows x 4212 columns]\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "with open(\"metadata_3837_genomes_june2016_complete.tsv\") as tsvfile:\n",
    "    reader = csv.DictReader(tsvfile, dialect = 'excel-tab')\n",
    "    for ind, row in enumerate(reader):\n",
    "        #if ind > 20: \n",
    "        #    break\n",
    "        genome_id = row['taxon_oid']\n",
    "        if genome_id in feature_matrix.index:\n",
    "            # convert the OD to a taxon\n",
    "            taxon = list(row.values())[4:-1]\n",
    "            feature_matrix['Y'] = feature_matrix['Y'].astype(object)\n",
    "            feature_matrix.at[genome_id, 'Y'] = taxon\n",
    "#print(feature_matrix.iloc[0:9,1])\n",
    "print(feature_matrix.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def genomicdistance(genome_id, reference_id):\n",
    "# Recursive algorithm which checks for similarity\n",
    "    def matching(taxon_1, taxon_2):\n",
    "        \n",
    "        if len(taxon_1) < 1 or len(taxon_2) < 1:\n",
    "            return 0\n",
    "        if taxon_1[0] == taxon_2[0]:\n",
    "            return 1 + matching(taxon_1[1:], taxon_2[1:])\n",
    "        else: \n",
    "            return 0\n",
    "  #print(reference_id in feature_matrix.index)\n",
    "    taxon_1 = feature_matrix.at[reference_id,'Y']\n",
    "    taxon_2 = feature_matrix.at[genome_id, 'Y']\n",
    "    #print(taxon_1)\n",
    "    #print(taxon_2)\n",
    "    if len(taxon_1) > len(taxon_2):\n",
    "        return matching(taxon_1, taxon_2) / len(taxon_1)\n",
    "    else:\n",
    "        return matching(taxon_1, taxon_2) / len(taxon_2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "genomes\n",
      "2576861342    0.000000\n",
      "2576861384    0.000000\n",
      "2576861345    0.000000\n",
      "2576861346    0.000000\n",
      "2576861348    0.000000\n",
      "2576861349    0.000000\n",
      "2576861368    0.000000\n",
      "2576861352    0.000000\n",
      "2576861377    0.000000\n",
      "2576861424    0.000000\n",
      "2576861435    0.000000\n",
      "2576861471    0.000000\n",
      "2576861456    0.833333\n",
      "2576861478    0.000000\n",
      "2576861425    0.000000\n",
      "2576861498    0.333333\n",
      "2576861499    0.333333\n",
      "2576861554    0.333333\n",
      "640427122     0.000000\n",
      "2576861487    0.000000\n",
      "2576861488    0.000000\n",
      "2576861656    0.000000\n",
      "2576861677    0.000000\n",
      "2576861679    0.000000\n",
      "2576861765    0.000000\n",
      "2576861768    0.000000\n",
      "2576861772    0.000000\n",
      "2576861745    0.000000\n",
      "2576861748    0.000000\n",
      "2576861752    0.000000\n",
      "2576861754    0.000000\n",
      "2576861755    0.000000\n",
      "2576861758    0.000000\n",
      "2576861779    0.000000\n",
      "2576861725    0.000000\n",
      "2576861726    0.000000\n",
      "2576861707    0.000000\n",
      "2576861731    0.000000\n",
      "2576861732    0.000000\n",
      "2576861737    0.000000\n",
      "2576861669    0.000000\n",
      "2576861740    0.000000\n",
      "2571042112    0.000000\n",
      "2571042114    0.000000\n",
      "2571042037    0.000000\n",
      "2571042294    0.000000\n",
      "2571042301    0.000000\n",
      "2571042144    0.000000\n",
      "2571042232    0.000000\n",
      "2571042147    0.000000\n",
      "Name: Y2, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "for genome, row in feature_matrix.iterrows():\n",
    "    #print(type(genome))\n",
    "    #print(genome)\n",
    "    distance = genomicdistance(genome, '2505679030')\n",
    "    feature_matrix.at[genome, 'Y2'] = distance\n",
    "    #print(distance)\n",
    "print(feature_matrix.iloc[450:500,-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_matrix.to_csv(\"plant_gene_matrix.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "genes       COG1522  COG0859  COG0483  COG1066  COG0034  COG2217  COG0101  \\\n",
      "genomes                                                                     \n",
      "2505679030      1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "651053067       1.0      0.0      1.0      1.0      1.0      1.0      1.0   \n",
      "637000240       1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "637000208       1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "650716076       1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "\n",
      "genes       COG0548  COG0642  COG1032  ...  COG1967  COG1148  COG3054  \\\n",
      "genomes                                ...                              \n",
      "2505679030      1.0      1.0      0.0  ...      0.0      0.0      0.0   \n",
      "651053067       1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "637000240       1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "637000208       1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "650716076       1.0      1.0      0.0  ...      0.0      0.0      0.0   \n",
      "\n",
      "genes       COG1153  COG4071  COG2214  COG5256  COG3388  \\\n",
      "genomes                                                   \n",
      "2505679030      0.0      0.0      0.0      0.0      0.0   \n",
      "651053067       0.0      0.0      0.0      0.0      0.0   \n",
      "637000240       0.0      0.0      0.0      0.0      0.0   \n",
      "637000208       0.0      0.0      0.0      0.0      0.0   \n",
      "650716076       0.0      0.0      0.0      0.0      0.0   \n",
      "\n",
      "genes                                                       Y   Y2  \n",
      "genomes                                                             \n",
      "2505679030  [Bacteroidetes, Bacteroidetes, Cytophagia, Cyt...  1.0  \n",
      "651053067   [Alphaproteobacteria, Proteobacteria, Alphapro...  0.0  \n",
      "637000240   [Alphaproteobacteria, Proteobacteria, Alphapro...  0.0  \n",
      "637000208   [Burkholderiales, Proteobacteria, Betaproteoba...  0.0  \n",
      "650716076   [Pseudomonas, Proteobacteria, Gammaproteobacte...  0.0  \n",
      "\n",
      "[5 rows x 4213 columns]\n"
     ]
    }
   ],
   "source": [
    "print(feature_matrix.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "genes       COG1522  COG0859  COG0483  COG1066  COG0034  COG2217  COG0101  \\\n",
      "genomes                                                                     \n",
      "2505679030      1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "651053067       1.0      0.0      1.0      1.0      1.0      1.0      1.0   \n",
      "637000240       1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "637000208       1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "650716076       1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "...             ...      ...      ...      ...      ...      ...      ...   \n",
      "2643221931      1.0      0.0      1.0      1.0      1.0      1.0      1.0   \n",
      "2643221933      1.0      0.0      1.0      1.0      1.0      1.0      1.0   \n",
      "2643221934      1.0      0.0      1.0      1.0      1.0      1.0      1.0   \n",
      "2643221904      1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "2643221912      1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "\n",
      "genes       COG0548  COG0642  COG1032  ...  COG1967  COG1148  COG3054  \\\n",
      "genomes                                ...                              \n",
      "2505679030      1.0      1.0      0.0  ...      0.0      0.0      0.0   \n",
      "651053067       1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "637000240       1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "637000208       1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "650716076       1.0      1.0      0.0  ...      0.0      0.0      0.0   \n",
      "...             ...      ...      ...  ...      ...      ...      ...   \n",
      "2643221931      1.0      1.0      0.0  ...      0.0      0.0      0.0   \n",
      "2643221933      1.0      1.0      0.0  ...      0.0      0.0      0.0   \n",
      "2643221934      1.0      1.0      0.0  ...      0.0      0.0      0.0   \n",
      "2643221904      1.0      1.0      0.0  ...      0.0      0.0      0.0   \n",
      "2643221912      1.0      1.0      0.0  ...      0.0      0.0      0.0   \n",
      "\n",
      "genes       COG1153  COG4071  COG2214  COG5256  COG3388  \\\n",
      "genomes                                                   \n",
      "2505679030      0.0      0.0      0.0      0.0      0.0   \n",
      "651053067       0.0      0.0      0.0      0.0      0.0   \n",
      "637000240       0.0      0.0      0.0      0.0      0.0   \n",
      "637000208       0.0      0.0      0.0      0.0      0.0   \n",
      "650716076       0.0      0.0      0.0      0.0      0.0   \n",
      "...             ...      ...      ...      ...      ...   \n",
      "2643221931      0.0      0.0      0.0      0.0      0.0   \n",
      "2643221933      0.0      0.0      0.0      1.0      0.0   \n",
      "2643221934      0.0      0.0      0.0      0.0      0.0   \n",
      "2643221904      0.0      0.0      0.0      0.0      0.0   \n",
      "2643221912      0.0      0.0      0.0      0.0      0.0   \n",
      "\n",
      "genes                                                       Y   Y2  \n",
      "genomes                                                             \n",
      "2505679030  [Bacteroidetes, Bacteroidetes, Cytophagia, Cyt...  1.0  \n",
      "651053067   [Alphaproteobacteria, Proteobacteria, Alphapro...  0.0  \n",
      "637000240   [Alphaproteobacteria, Proteobacteria, Alphapro...  0.0  \n",
      "637000208   [Burkholderiales, Proteobacteria, Betaproteoba...  0.0  \n",
      "650716076   [Pseudomonas, Proteobacteria, Gammaproteobacte...  0.0  \n",
      "...                                                       ...  ...  \n",
      "2643221931  [Alphaproteobacteria, Proteobacteria, Alphapro...  0.0  \n",
      "2643221933  [Alphaproteobacteria, Proteobacteria, Alphapro...  0.0  \n",
      "2643221934  [Alphaproteobacteria, Proteobacteria, Alphapro...  0.0  \n",
      "2643221904  [Actinobacteria_one, Actinobacteria, Actinobac...  0.0  \n",
      "2643221912  [Actinobacteria_one, Actinobacteria, Actinobac...  0.0  \n",
      "\n",
      "[3837 rows x 4213 columns]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "Y = pd.DataFrame(data = feature_matrix.iloc[:, -1], index = feature_matrix.index)\n",
    "feature_matrix.drop(columns = ['Y', 'Y2'])\n",
    "print(feature_matrix)\n",
    "\n",
    "#print(Y)\n",
    "from sklearn.datasets import load_iris\n",
    "from sklearn.feature_selection import SelectKBest\n",
    "from sklearn.feature_selection import chi2\n",
    "X, y = load_iris(return_X_y=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Originally had  4211  features\n",
      "genomes\n",
      "2505679030    1.0\n",
      "651053067     0.0\n",
      "637000240     0.0\n",
      "637000208     0.0\n",
      "650716076     0.0\n",
      "             ... \n",
      "2643221931    0.0\n",
      "2643221933    0.0\n",
      "2643221934    0.0\n",
      "2643221904    0.0\n",
      "2643221912    0.0\n",
      "Name: Y2, Length: 3837, dtype: float64\n",
      "genes       COG0859  COG0483  COG1032  COG0559  COG1804  COG1492  COG3173  \\\n",
      "genomes                                                                     \n",
      "2505679030      1.0      1.0      0.0      0.0      1.0      0.0      1.0   \n",
      "651053067       0.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "637000240       1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "637000208       1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "650716076       1.0      1.0      0.0      1.0      1.0      1.0      1.0   \n",
      "...             ...      ...      ...      ...      ...      ...      ...   \n",
      "2643221931      0.0      1.0      0.0      1.0      1.0      1.0      0.0   \n",
      "2643221933      0.0      1.0      0.0      1.0      1.0      1.0      0.0   \n",
      "2643221934      0.0      1.0      0.0      1.0      1.0      1.0      0.0   \n",
      "2643221904      1.0      1.0      0.0      1.0      0.0      0.0      1.0   \n",
      "2643221912      1.0      1.0      0.0      1.0      0.0      0.0      1.0   \n",
      "\n",
      "genes       COG0625  COG1638  COG1523  ...  COG4283  COG4245  COG3529  \\\n",
      "genomes                                ...                              \n",
      "2505679030      0.0      0.0      0.0  ...      0.0      0.0      0.0   \n",
      "651053067       1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "637000240       1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "637000208       1.0      1.0      0.0  ...      0.0      0.0      0.0   \n",
      "650716076       1.0      0.0      1.0  ...      0.0      0.0      1.0   \n",
      "...             ...      ...      ...  ...      ...      ...      ...   \n",
      "2643221931      1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "2643221933      1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "2643221934      1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "2643221904      0.0      0.0      1.0  ...      0.0      0.0      0.0   \n",
      "2643221912      0.0      0.0      1.0  ...      0.0      0.0      0.0   \n",
      "\n",
      "genes       COG3199  COG3410  COG3059  COG3076  COG3160  COG1913  COG2456  \n",
      "genomes                                                                    \n",
      "2505679030      0.0      0.0      0.0      0.0      0.0      0.0      0.0  \n",
      "651053067       0.0      0.0      0.0      0.0      0.0      0.0      0.0  \n",
      "637000240       0.0      0.0      0.0      0.0      0.0      0.0      0.0  \n",
      "637000208       0.0      0.0      0.0      0.0      0.0      0.0      0.0  \n",
      "650716076       1.0      0.0      0.0      0.0      1.0      0.0      0.0  \n",
      "...             ...      ...      ...      ...      ...      ...      ...  \n",
      "2643221931      0.0      0.0      0.0      0.0      0.0      0.0      0.0  \n",
      "2643221933      0.0      0.0      0.0      0.0      0.0      0.0      0.0  \n",
      "2643221934      0.0      0.0      0.0      0.0      0.0      0.0      0.0  \n",
      "2643221904      0.0      0.0      0.0      0.0      0.0      0.0      0.0  \n",
      "2643221912      0.0      0.0      0.0      0.0      0.0      0.0      0.0  \n",
      "\n",
      "[3837 rows x 3098 columns]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#print(feature_matrix)\n",
    "\n",
    "# split feature matrix into features and target\n",
    "X = feature_matrix.iloc[:, 0:-2]\n",
    "#print(X)\n",
    "Y = feature_matrix.iloc[:, -1]\n",
    "#print(Y)\n",
    "print(\"Originally had \", str(len(X.columns)), ' features')\n",
    "\n",
    "# remove features with low variance, aka features with low correlation to a microbe's function/taxonomy\n",
    "selector = VarianceThreshold(threshold=(.05)).fit(X,Y)\n",
    "\n",
    "# selecting only most salient features,alternatively could use df.drop()\n",
    "vdf = feature_matrix[feature_matrix.columns[selector.get_support(indices=True)]]\n",
    "\n",
    "#print('V-based reduced feature matrix with ', str(len(vdf.columns)), 'features: \\n', list(vdf.columns))\n",
    "#print('\\n from ', str(len(vdf)), ' genomes')\n",
    "\n",
    "# update the target\n",
    "Y = feature_matrix.loc[list(vdf.index), 'Y2']\n",
    "print(Y)\n",
    "\n",
    "X = vdf\n",
    "print(X)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the 10 best features correlated to genomiic distance?\n",
    "#bestfeatures = SelectKBest(k=100)\n",
    "#fit = bestfeatures.fit(X,Y)\n",
    "#dfscores = pd.DataFrame(fit.scores_)\n",
    "#dfcolumns = pd.DataFrame(X.columns)\n",
    "\n",
    "#concat two dataframes for better visualization \n",
    "#featureScores = pd.concat([dfcolumns,dfscores],axis=1)\n",
    "#featureScores.columns = ['Specs','Score']  #naming the dataframe columns\n",
    "#print(featureScores.nlargest(100,'Score'))  #print 10 best features\n",
    "#print(type(featureScores))\n",
    "#print(featureScores)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'vdf' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-1-7959f8c4d7a6>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[1;31m#feature_matrix = pd.read_csv(\"plant_gene_matrix.csv\")\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[1;31m#print(feature_matrix)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 5\u001b[1;33m \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mvdf\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m: name 'vdf' is not defined"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "import pandas as pd\n",
    "#feature_matrix = pd.read_csv(\"plant_gene_matrix.csv\")\n",
    "#print(feature_matrix)\n",
    "print(vdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "genes       COG0859  COG0483  COG1032  COG0559  COG1804  COG1492  COG3173  \\\n",
      "genomes                                                                     \n",
      "2505679030      1.0      1.0      0.0      0.0      1.0      0.0      1.0   \n",
      "651053067       0.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "637000240       1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "637000208       1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "650716076       1.0      1.0      0.0      1.0      1.0      1.0      1.0   \n",
      "\n",
      "genes       COG0625  COG1638  COG1523  ...  COG4283  COG4245  COG3529  \\\n",
      "genomes                                ...                              \n",
      "2505679030      0.0      0.0      0.0  ...      0.0      0.0      0.0   \n",
      "651053067       1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "637000240       1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "637000208       1.0      1.0      0.0  ...      0.0      0.0      0.0   \n",
      "650716076       1.0      0.0      1.0  ...      0.0      0.0      1.0   \n",
      "\n",
      "genes       COG3199  COG3410  COG3059  COG3076  COG3160  COG1913  COG2456  \n",
      "genomes                                                                    \n",
      "2505679030      0.0      0.0      0.0      0.0      0.0      0.0      0.0  \n",
      "651053067       0.0      0.0      0.0      0.0      0.0      0.0      0.0  \n",
      "637000240       0.0      0.0      0.0      0.0      0.0      0.0      0.0  \n",
      "637000208       0.0      0.0      0.0      0.0      0.0      0.0      0.0  \n",
      "650716076       1.0      0.0      0.0      0.0      1.0      0.0      0.0  \n",
      "\n",
      "[5 rows x 3098 columns]\n"
     ]
    }
   ],
   "source": [
    "print(vdf.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Correlation Matrix? yikes.\n",
    "# If doing correlation, make sure that 'Y' is included in df\n",
    "#import matplotlib.pyplot as plt\n",
    "#import seaborn as sns\n",
    "#Using Pearson Correlation\n",
    "#plt.figure(figsize=(12,10))\n",
    "#cor = feature_matrix.corr()\n",
    "#sns.heatmap(cor, annot=True, cmap=plt.cm.Reds)\n",
    "#plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#UMAP time\n",
    "\n",
    "import umap.umap_ as umap\n",
    "reducer = umap.UMAP()\n",
    "embedding = reducer.fit_transform(vdf)\n",
    "embedding.shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
