{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Introduction\n",
    "\n",
    "This project is based off data of 3867 annotated bacterial genomes with their corresponding genes in the form of COG ids, provided by Levy et. al. By clustering these genomes together based on taxonomic similarity, we hope to obtain groups of highly-related bacteria. Of these, we are interested in plant-associated bacteria which may be responsible for symbiotic relationships with plants, namely agricultural crops. By obtaining the genes unique to these plant-associated bacteria, we identify genes of interest that may be responsible for the formation of bacterial natural products that defer advantages for agricultural crops in the form of pest resistance, without the need for commercial pesticides.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import block\n",
    "import umap\n",
    "import umap.plot\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import hdbscan\n",
    "import bokeh\n",
    "\n",
    "from sklearn.feature_selection import SelectKBest\n",
    "from sklearn.feature_selection import chi2\n",
    "from sklearn.feature_selection import VarianceThreshold"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Featurisation \n",
    "The file containing all the COG ids is formatted such that each line contains 2 important fields: the \n",
    "IMG id (corresponding to a certain bacterial genome), and the COG ID, a single gene in that matrix. In order to featurize all our COG IDs, we create a function that returns all the unique genomes (IMG ids) and genes (COG IDs) by parsing the COGfile."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Constants/filenames\n",
    "gene_index = 4\n",
    "genome_index = 1\n",
    "cogfile = '/home/ana/wet_lab_common/project-plant-genes/cogs.txt'\n",
    "\n",
    "def genomes_genes(file, genome_index, gene_index):\n",
    "    unique_genomes = []\n",
    "    unique_genes = []\n",
    "    with open(file, 'r+') as fh:\n",
    "        \n",
    "        for line in fh.readlines()[1:]:\n",
    "            if line.split()[genome_index] not in unique_genomes:\n",
    "                unique_genomes.append(line.split()[genome_index])\n",
    "            if line.split()[gene_index] not in unique_genes:\n",
    "                unique_genes.append(line.split()[gene_index])\n",
    "    return [unique_genomes, unique_genes]\n",
    "allgenomes, allgenes = genomes_genes(cogfile, genome_index, gene_index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We then construct a Pandas dataframe of zeros where the indices consist of the genomes, and the column names consist of each gene. By parsing the COGfile a second time, one can read each genome annotation line by line. Concurrently, one  annotates the matrix with 1s for the entry represented by each line, taking the IMG id as the index, and the COG id as the column index. By this process, a feature matrix showing all bacterial genomes, with the presence or absence of all unique genes as “features”, is created. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def matrix(file, genes, genomes, gene_col, genome_col):\n",
    "    gcols = pd.Index(genes, name = \"genes\")\n",
    "    grows = pd.Index(genomes, name = \"genomes\")\n",
    "    df = pd.DataFrame(data = np.zeros(shape = (len(genomes), len(genes))), index = grows, columns = gcols)\n",
    "    fh = open(file, 'r+')\n",
    "    for line in fh.readlines()[1:]:       \n",
    "        genome = line.split()[genome_index]\n",
    "        gene = line.split()[gene_index]\n",
    "        # mark gene as present\n",
    "        df.at[genome, gene] = 1\n",
    "    fh.close()\n",
    "    return df\n",
    "feature_matrix = matrix(cogfile, allgenes, allgenomes, gene_index, genome_index)\n",
    "feature_matrix['taxon'] = [0] * len(feature_matrix)\n",
    "print(feature_matrix.head(10))\n",
    "print(len(feature_matrix))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, one must parse the metadata in order to obtain the taxonomy, association (plant, nonplant, soil), and isolation site of each bacteria. The metadata comes in the form of a TSV with 1 line for each genome. It is read line by line, and a secondary matrix is created concurrently with the columns ‘taxon’, ‘association’, and ‘site’, indexed by each genome ID.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "genes       COG1522  COG0859  COG0483  COG1066  COG0034  COG2217  COG0101  \\\n",
      "genomes                                                                     \n",
      "2505679030      1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "651053067       1.0      0.0      1.0      1.0      1.0      1.0      1.0   \n",
      "637000240       1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "637000208       1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "650716076       1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "\n",
      "genes       COG0548  COG0642  COG1032  ...  COG1148  COG3054  COG1153  \\\n",
      "genomes                                ...                              \n",
      "2505679030      1.0      1.0      0.0  ...      0.0      0.0      0.0   \n",
      "651053067       1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "637000240       1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "637000208       1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "650716076       1.0      1.0      0.0  ...      0.0      0.0      0.0   \n",
      "\n",
      "genes       COG4071  COG2214  COG5256  COG3388  \\\n",
      "genomes                                          \n",
      "2505679030      0.0      0.0      0.0      0.0   \n",
      "651053067       0.0      0.0      0.0      0.0   \n",
      "637000240       0.0      0.0      0.0      0.0   \n",
      "637000208       0.0      0.0      0.0      0.0   \n",
      "650716076       0.0      0.0      0.0      0.0   \n",
      "\n",
      "genes                                                   taxon     site  \\\n",
      "genomes                                                                  \n",
      "2505679030  [Bacteroidetes, Bacteroidetes, Cytophagia, Cyt...  Unknown   \n",
      "651053067   [Alphaproteobacteria, Proteobacteria, Alphapro...  Unknown   \n",
      "637000240   [Alphaproteobacteria, Proteobacteria, Alphapro...  Unknown   \n",
      "637000208   [Burkholderiales, Proteobacteria, Betaproteoba...  Unknown   \n",
      "650716076   [Pseudomonas, Proteobacteria, Gammaproteobacte...  Unknown   \n",
      "\n",
      "genes       Classification  \n",
      "genomes                     \n",
      "2505679030             NPA  \n",
      "651053067               PA  \n",
      "637000240              NPA  \n",
      "637000208              NPA  \n",
      "650716076             soil  \n",
      "\n",
      "[5 rows x 4214 columns]\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "with open(\"/home/ana/wet_lab_common/project-plant-genes/metadata_3837_genomes_june2016_complete.tsv\") as tsvfile:\n",
    "    reader = csv.DictReader(tsvfile, dialect = 'excel-tab')\n",
    "    feature_matrix['site'] = [0] * len(feature_matrix)\n",
    "    feature_matrix['Classification'] = [0] * len(feature_matrix)\n",
    "    for ind, row in enumerate(reader):\n",
    "        #if ind > 20: \n",
    "        #    break\n",
    "        genome_id = row['taxon_oid']\n",
    "        site = row['Isolation_Site']\n",
    "        association = row['Classification']\n",
    "        if genome_id in feature_matrix.index:\n",
    "            # convert the OD to a taxon\n",
    "            taxon = list(row.values())[4:-1]\n",
    "            feature_matrix['taxon'] = feature_matrix['taxon'].astype(object)\n",
    "            feature_matrix.at[genome_id, 'taxon'] = taxon\n",
    "            feature_matrix['site'] = feature_matrix['site'].astype(object)\n",
    "            feature_matrix.at[genome_id, 'site'] = site\n",
    "            feature_matrix['Classification'] = feature_matrix['Classification'].astype(object)\n",
    "            feature_matrix.at[genome_id, 'Classification'] = association\n",
    "print(feature_matrix.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we create a function to define the genomic similarity between 2 bacterial genomes. This will be important in validating the distance between points in our clusters. This function takes in 2 genome IDs, a reference and a test. It searches the second matrix for the taxa of each genome, and finds the difference between the 2 taxa based on a Bray-Curtis similarity metric (0 means completely dissimilar, 1 means identical)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [],
   "source": [
    "def genomicdistance(genome_id, reference_id):\n",
    "# Recursive algorithm which checks for similarity\n",
    "    def matching(taxon_1, taxon_2):\n",
    "        \n",
    "        if len(taxon_1) < 1 or len(taxon_2) < 1:\n",
    "            return 0\n",
    "        if taxon_1[0] == taxon_2[0]:\n",
    "            return 1 + matching(taxon_1[1:], taxon_2[1:])\n",
    "        else: \n",
    "            return 0\n",
    "    taxon_1 = feature_matrix.at[reference_id,'taxon']\n",
    "    taxon_2 = feature_matrix.at[genome_id, 'taxon']\n",
    "    if len(taxon_1) > len(taxon_2):\n",
    "        return matching(taxon_1, taxon_2) / len(taxon_1)\n",
    "    else:\n",
    "        return matching(taxon_1, taxon_2) / len(taxon_2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Creating a dataframe of the distances:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "genomes\n",
      "2576861342    0.000000\n",
      "2576861384    0.000000\n",
      "2576861345    0.000000\n",
      "2576861346    0.000000\n",
      "2576861348    0.000000\n",
      "2576861349    0.000000\n",
      "2576861368    0.000000\n",
      "2576861352    0.000000\n",
      "2576861377    0.000000\n",
      "2576861424    0.000000\n",
      "2576861435    0.000000\n",
      "2576861471    0.000000\n",
      "2576861456    0.833333\n",
      "2576861478    0.000000\n",
      "2576861425    0.000000\n",
      "2576861498    0.333333\n",
      "2576861499    0.333333\n",
      "2576861554    0.333333\n",
      "640427122     0.000000\n",
      "2576861487    0.000000\n",
      "2576861488    0.000000\n",
      "2576861656    0.000000\n",
      "2576861677    0.000000\n",
      "2576861679    0.000000\n",
      "2576861765    0.000000\n",
      "2576861768    0.000000\n",
      "2576861772    0.000000\n",
      "2576861745    0.000000\n",
      "2576861748    0.000000\n",
      "2576861752    0.000000\n",
      "2576861754    0.000000\n",
      "2576861755    0.000000\n",
      "2576861758    0.000000\n",
      "2576861779    0.000000\n",
      "2576861725    0.000000\n",
      "2576861726    0.000000\n",
      "2576861707    0.000000\n",
      "2576861731    0.000000\n",
      "2576861732    0.000000\n",
      "2576861737    0.000000\n",
      "2576861669    0.000000\n",
      "2576861740    0.000000\n",
      "2571042112    0.000000\n",
      "2571042114    0.000000\n",
      "2571042037    0.000000\n",
      "2571042294    0.000000\n",
      "2571042301    0.000000\n",
      "2571042144    0.000000\n",
      "2571042232    0.000000\n",
      "2571042147    0.000000\n",
      "Name: distance, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "for genome, row in feature_matrix.iterrows():\n",
    "    distance = genomicdistance(genome, '2505679030')\n",
    "    feature_matrix.at[genome, 'distance'] = distance\n",
    "print(feature_matrix.iloc[450:500,-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "genes       COG1522  COG0859  COG0483  COG1066  COG0034  COG2217  COG0101  \\\n",
      "genomes                                                                     \n",
      "2505679030      1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "651053067       1.0      0.0      1.0      1.0      1.0      1.0      1.0   \n",
      "637000240       1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "637000208       1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "650716076       1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "\n",
      "genes       COG0548  COG0642  COG1032  ...  COG3054  COG1153  COG4071  \\\n",
      "genomes                                ...                              \n",
      "2505679030      1.0      1.0      0.0  ...      0.0      0.0      0.0   \n",
      "651053067       1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "637000240       1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "637000208       1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "650716076       1.0      1.0      0.0  ...      0.0      0.0      0.0   \n",
      "\n",
      "genes       COG2214  COG5256  COG3388  \\\n",
      "genomes                                 \n",
      "2505679030      0.0      0.0      0.0   \n",
      "651053067       0.0      0.0      0.0   \n",
      "637000240       0.0      0.0      0.0   \n",
      "637000208       0.0      0.0      0.0   \n",
      "650716076       0.0      0.0      0.0   \n",
      "\n",
      "genes                                                   taxon     site  \\\n",
      "genomes                                                                  \n",
      "2505679030  [Bacteroidetes, Bacteroidetes, Cytophagia, Cyt...  Unknown   \n",
      "651053067   [Alphaproteobacteria, Proteobacteria, Alphapro...  Unknown   \n",
      "637000240   [Alphaproteobacteria, Proteobacteria, Alphapro...  Unknown   \n",
      "637000208   [Burkholderiales, Proteobacteria, Betaproteoba...  Unknown   \n",
      "650716076   [Pseudomonas, Proteobacteria, Gammaproteobacte...  Unknown   \n",
      "\n",
      "genes       Classification  distance  \n",
      "genomes                               \n",
      "2505679030             NPA       1.0  \n",
      "651053067               PA       0.0  \n",
      "637000240              NPA       0.0  \n",
      "637000208              NPA       0.0  \n",
      "650716076             soil       0.0  \n",
      "\n",
      "[5 rows x 4215 columns]\n"
     ]
    }
   ],
   "source": [
    "print(feature_matrix.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Having finished our preprocessing, we move on to perform feature selection on our dataset in order to remove genes that are not indicative of taxonomic differences. In other words, core metabolic genes that may be shared among all bacteria. Using scikit-learn’s VarianceThreshold, we take our feature matrix and remove from it any genes (‘features’) which have a variance less than 0.1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import feature selection package\n",
    "from sklearn.feature_selection import VarianceThreshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Originally had  4211  features\n",
      "genomes\n",
      "2505679030    1.0\n",
      "651053067     0.0\n",
      "637000240     0.0\n",
      "637000208     0.0\n",
      "650716076     0.0\n",
      "             ... \n",
      "2643221931    0.0\n",
      "2643221933    0.0\n",
      "2643221934    0.0\n",
      "2643221904    0.0\n",
      "2643221912    0.0\n",
      "Name: distance, Length: 3837, dtype: float64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/nathan/miniconda3/envs/jhub/lib/python3.6/site-packages/ipykernel_launcher.py:22: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "genes       COG0859  COG1032  COG0559  COG1804  COG1492  COG3173  COG0625  \\\n",
      "genomes                                                                     \n",
      "2505679030      1.0      0.0      0.0      1.0      0.0      1.0      0.0   \n",
      "651053067       0.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "637000240       1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "637000208       1.0      1.0      1.0      1.0      1.0      1.0      1.0   \n",
      "650716076       1.0      0.0      1.0      1.0      1.0      1.0      1.0   \n",
      "...             ...      ...      ...      ...      ...      ...      ...   \n",
      "2643221931      0.0      0.0      1.0      1.0      1.0      0.0      1.0   \n",
      "2643221933      0.0      0.0      1.0      1.0      1.0      0.0      1.0   \n",
      "2643221934      0.0      0.0      1.0      1.0      1.0      0.0      1.0   \n",
      "2643221904      1.0      0.0      1.0      0.0      0.0      1.0      0.0   \n",
      "2643221912      1.0      0.0      1.0      0.0      0.0      1.0      0.0   \n",
      "\n",
      "genes       COG1638  COG1523  COG3959  ...  COG4324  COG4825  COG1938  \\\n",
      "genomes                                ...                              \n",
      "2505679030      0.0      0.0      1.0  ...      1.0      0.0      0.0   \n",
      "651053067       1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "637000240       1.0      1.0      0.0  ...      0.0      0.0      0.0   \n",
      "637000208       1.0      0.0      1.0  ...      1.0      0.0      0.0   \n",
      "650716076       0.0      1.0      1.0  ...      1.0      0.0      0.0   \n",
      "...             ...      ...      ...  ...      ...      ...      ...   \n",
      "2643221931      1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "2643221933      1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "2643221934      1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
      "2643221904      0.0      1.0      0.0  ...      0.0      0.0      0.0   \n",
      "2643221912      0.0      1.0      0.0  ...      0.0      0.0      1.0   \n",
      "\n",
      "genes       COG4982  COG0615  COG4840  COG1166  COG3529  COG3059  association  \n",
      "genomes                                                                        \n",
      "2505679030      0.0      0.0      0.0      0.0      0.0      0.0          NPA  \n",
      "651053067       0.0      0.0      0.0      0.0      0.0      0.0           PA  \n",
      "637000240       0.0      0.0      0.0      0.0      0.0      0.0          NPA  \n",
      "637000208       0.0      0.0      0.0      0.0      0.0      0.0          NPA  \n",
      "650716076       0.0      0.0      0.0      1.0      1.0      0.0         soil  \n",
      "...             ...      ...      ...      ...      ...      ...          ...  \n",
      "2643221931      0.0      0.0      0.0      0.0      0.0      0.0           PA  \n",
      "2643221933      0.0      0.0      0.0      0.0      0.0      0.0           PA  \n",
      "2643221934      0.0      0.0      0.0      0.0      0.0      0.0           PA  \n",
      "2643221904      0.0      0.0      0.0      0.0      0.0      0.0           PA  \n",
      "2643221912      0.0      0.0      0.0      0.0      0.0      0.0           PA  \n",
      "\n",
      "[3837 rows x 2519 columns]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# split feature matrix into features and target\n",
    "X = feature_matrix.iloc[:, 0:-4]\n",
    "\n",
    "#print(distance)\n",
    "print(\"Originally had \", str(len(X.columns)), ' features')\n",
    "\n",
    "# remove features with low variance, aka features with low correlation to a microbe's function/taxonomy\n",
    "selector = VarianceThreshold(threshold=(.1)).fit(X)\n",
    "\n",
    "# selecting only most salient features,alternatively could use df.drop()\n",
    "vdf = feature_matrix[feature_matrix.columns[selector.get_support(indices=True)]]\n",
    "\n",
    "#print('V-based reduced feature matrix with ', str(len(vdf.columns)), 'features: \\n', list(vdf.columns))\n",
    "#print('\\n from ', str(len(vdf)), ' genomes')\n",
    "\n",
    "# update the target\n",
    "distance = feature_matrix.loc[list(vdf.index), 'distance']\n",
    "print(distance)\n",
    "site = feature_matrix.loc[list(vdf.index), 'site']\n",
    "taxon = feature_matrix.loc[list(vdf.index), 'taxon']\n",
    "association = feature_matrix.loc[list(vdf.index), 'Classification']\n",
    "vdf['association'] =association\n",
    "vdf.to_pickle('feature_matrix.pkl')\n",
    "X = vdf\n",
    "print(X)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note:\n",
    "     The threshold of 0.1 is somewhat arbitrary. This is due to the difficulty in estimating the variance in genes of interest as we cannot know exactly which genes are responsible for plant association/natural product production until validation in the wet lab. However, a variance threshold of 0.1 seemed to produce good clusters (explained later) while reducing the number of features by a sizeable one third, so we stuck with it."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, in order to prepare our data for clustering, we reduce the dimensions of the data using UMAP-Uniform Manifold Approximation and Projection. After importing the package, we feed in our data, choosing ‘braycurtis’ as the metric. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-7eba1ad9ecdc>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32mimport\u001b[0m \u001b[0mumap\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mumap_\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mumap\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mmapper\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mumap\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mUMAP\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmetric\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'braycurtis'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn_neighbors\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m30\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmin_dist\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m0.0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvdf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0membedding\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmapper\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtransform\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvdf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0membedding\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/nathan/miniconda3/envs/jhub/lib/python3.6/site-packages/umap/__init__.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mumap_\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mUMAP\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;31m# Workaround: https://github.com/numba/numba/issues/3341\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mnumba\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/nathan/miniconda3/envs/jhub/lib/python3.6/site-packages/umap/umap_.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     24\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mscipy\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msparse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     25\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mscipy\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msparse\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcsgraph\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 26\u001b[0;31m \u001b[0;32mimport\u001b[0m \u001b[0mnumba\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     27\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     28\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mumap\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdistances\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mdist\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/nathan/miniconda3/envs/jhub/lib/python3.6/site-packages/numba/__init__.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     13\u001b[0m \u001b[0;32mdel\u001b[0m \u001b[0mget_versions\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 15\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0mnumba\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcore\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mconfig\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     16\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mnumba\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtesting\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0m_runtests\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mruntests\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mnumba\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcore\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mtypes\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/nathan/miniconda3/envs/jhub/lib/python3.6/site-packages/numba/core/config.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 16\u001b[0;31m \u001b[0;32mimport\u001b[0m \u001b[0mllvmlite\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbinding\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mll\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     17\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[0mIS_WIN32\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mplatform\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstartswith\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'win32'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/nathan/miniconda3/envs/jhub/lib/python3.6/site-packages/llvmlite/binding/__init__.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mtransforms\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mvalue\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 14\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0manalysis\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     15\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mobject_file\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mcontext\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/nathan/miniconda3/envs/jhub/lib/python3.6/site-packages/llvmlite/binding/analysis.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mctypes\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mPOINTER\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mc_char_p\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mc_int\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0mllvmlite\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mir\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      8\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mllvmlite\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbinding\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mffi\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mllvmlite\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbinding\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmodule\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mparse_assembly\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/nathan/miniconda3/envs/jhub/lib/python3.6/site-packages/llvmlite/ir/__init__.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mmodule\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mbuilder\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      9\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0minstructions\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mtransforms\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/nathan/miniconda3/envs/jhub/lib/python3.6/importlib/_bootstrap.py\u001b[0m in \u001b[0;36m_find_and_load\u001b[0;34m(name, import_)\u001b[0m\n",
      "\u001b[0;32m/home/nathan/miniconda3/envs/jhub/lib/python3.6/importlib/_bootstrap.py\u001b[0m in \u001b[0;36m_find_and_load_unlocked\u001b[0;34m(name, import_)\u001b[0m\n",
      "\u001b[0;32m/home/nathan/miniconda3/envs/jhub/lib/python3.6/importlib/_bootstrap.py\u001b[0m in \u001b[0;36m_load_unlocked\u001b[0;34m(spec)\u001b[0m\n",
      "\u001b[0;32m/home/nathan/miniconda3/envs/jhub/lib/python3.6/importlib/_bootstrap_external.py\u001b[0m in \u001b[0;36mexec_module\u001b[0;34m(self, module)\u001b[0m\n",
      "\u001b[0;32m/home/nathan/miniconda3/envs/jhub/lib/python3.6/importlib/_bootstrap_external.py\u001b[0m in \u001b[0;36mget_code\u001b[0;34m(self, fullname)\u001b[0m\n",
      "\u001b[0;32m/home/nathan/miniconda3/envs/jhub/lib/python3.6/importlib/_bootstrap_external.py\u001b[0m in \u001b[0;36mget_data\u001b[0;34m(self, path)\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import umap.umap_ as umap\n",
    "\n",
    "mapper = umap.UMAP(metric = 'braycurtis', n_neighbors = 30, min_dist = 0.0).fit(vdf)\n",
    "embedding = mapper.transform(vdf)\n",
    "print(embedding.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Our returned embedding is then in two dimensions, which we can plot using matplotlib to get a rough idea of what the clustering might look like."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nWhere a and b are assumed to be 2 points in the UMAP space\\n\\n'"
      ]
     },
     "execution_count": 225,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjwAAAERCAYAAAB/1X3vAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzt3Xu8VHW9//HXh6sChqCIN3AbGAaaptu8KybmtUjN0nOO6al+pnaOZh4Tb0dKTcqMstIy89bxaGqaJmqKCVpeNyYmCidQEG+wFUTdyP3z++O71p61h5nZszezZ82s/X4+HvOYmXWbz3zXZT7zXd/1XebuiIiIiGRZj7QDEBEREelqSnhEREQk85TwiIiISOYp4REREZHMU8IjIiIimaeER0RERDJPCU83ZWYePRrSjqUcZjY/inds2rFUg5ntYGa3mdnbZrY2+u43ljFfQ7xuqxBmZpjZjVG5TazgMsdGy5xfqWVKtmmb6VplJTyJH5uJGzJt4kfWzezeMpb1x+Q8ZUzf28wWR9OvNbNhHYg3+VhrZu+a2eNmdpaZ9WtvOdI5ZrarmU00s5PTjqVWmNlg4HHgK8BQYCmwCFiWZlxdwcxOjtb/rmnHIpJV0T420cw2TTuWDRUlhRPN7IsdnTfNGp7DzGxIsZFmthlwRAeXeQQQL7MH8G8dmLeF8KMS/7AMBvYDfgI8a2ZbdDCWWjcneqxOOY5dgYuBk9uZbh4h3uVdHVANOIGQ6PwfsLW7b+7uW7r7mSnH1RVOJqz/tBOetwjb1zspxyHSFS6OHnWf8ABjCd+lbhKe14DehAN7MSdE0yzowHJPip5/k/e+HD+OflS2dPfBwObAZYADo4FrO7CsmufuO0aPN9KOpRzufnAU7zNpx1IFY6LnP7n7W6lG0k24+3nR9vWLtGMRka6RVsLzP9HziSWm+WretCVFNUJHAiuAc4BXgFFmtmdnAnT3d939QuCGaNB4M9u6M8sS6aCNo+cPU41CRCRL3L3dBzCfUNMxcUOmjYY7cDghIXHgkwWmGxWNWwAcEs/Xzuf+RzTdHdH7S6L3V2/IdwPGJeI+qpzyyvuuDcBOwG3A24SEbDZwEdC3yLzTonlPJlRB/jCaZznwXoHpjwEeBJqBlcDrwC3AbuXEV2T8AOB84FnCKb4VwD+Bq4Bh7Xz3YcCVwIvAB9HjJeC3wEEFYij2GFtgPY0t8plDo8+My2kZ8AxwdolyvjFe90BP4NvAzGj+JcB9QGO567zIZ5S9bhLrveCjzM9rSE4P7Bt9j+boez1P2Fd6FJl/MKFm9A9RWX5AON37EuH07tZlxHAocGf0XVcStvungAvjbSfatkut+/kFltsniv3xaP2sJBwjrqfAcaTAOu4LXAC8EH0vBzbNn67AMj4ebUePAK8S9oX3ou90NrBxkc8eW+y7lLkuBwGTCdv+SmAhcB1h/2p32YTjzvV5Mf8NOBXoXca2U+y41afEZ/YAvg5Mj9bRiujzrwVGllNO0fYzNZr/PeBhYO/E9AMJte//B3wUlcsPi62HxHz7Rd8n3i7fjT7nBMCKzLM9cE3is5ZH29w04Dxg806s14HAjxPrZSHhrMS2pdYrndg3E9t1sceNefvXkVEsMwmnd1dE3/cWYPcS36kPcCbwRLTOVhOah8wEfplcf53dp5PbZ4lHQ8myL3MFzaeyCc9hwPej15cXmO4H0bjLSCQc7Xzus9F0R0fvd4zeL6HID1453w34ZCLuf+nARt06D+GfuhN+hFcmxj0JDCgw77Ro/DmEtisebXjvk0h4CAeXmxLLW0No4Bq/Xwuc1k58620g0Xeen5hmdeI7xGW6b5HlHks4KMTTfkTux6XNjkw4kC6Lhq+K3icf+xRYT2MLfOZnCAev+DPejz43fv88sEWJg8GlwAOJOJLxfkSRnbWd9d/hdQPcFX3vOPYPk+VR5uc2JD7j2GjdefTZqxPj7gZ6FZj/x4lp4m12TeL9YuBTRT67D/C7vPnfy/vcidG0X4m+16rE5yTX/bN5y94qWo/J8ns/bz0dU2IdTwKeTqzj9yg/4WlKfM66qCzXJYY9C2xSYL6xdDLhIfz4vZr4jOWJ7XIx8I1Syyb8iKxNzP9h3np8FOhXYtv5HLn9+L28Zf2xyGf2A/6cmC5ZzvE6Gl+qnIDTo7JdS+7YEM+7H6GN5j8S3yl5PL2vRHn+MDFdfIxIfqdbyfsTAOyWt42tou0+7MBhHVyvWxH+OBY6Pi4mJIsF1yud2DeBnxH2p3iaZtruZz9LTHtU3vJbaHscXQ2cWCCuXrT9sxbvI8nYbitSFmXv04RE/21yv0Ufsf5vRuk/42WupPmUSArKnTa5kQAjo9evJTc0wAgZnhNqetpNeAhtbOKDet/E8BnR8C919rsR/m3EcR/RgQ07ecB/Btg5Gt6H8O82PphcW2DeeOP5ICqfw+IyIvEvCZiQ2MAuJDroAtsAtyc2ogNKxNeQN3wguQPt3cCniX4YCQfEm6NxbxP9WCTm3Zvcj9tfgD2I/jkRDlRfBK7Pm+fkaPppZW5XY/OGDwLejMa9AOwRDe8JfImQnDnwcIFl3pjYbt4Fvkz07xX4FLkD6zPlrvcKrZs4rnb3twLzNuRtew8A20fj+hOS6PhAf36B+c8CLo/W+4BEWe5OqKlyQs3dev+ICf/knHCgmwgMjYb3AnYA/gs4pci2fnKJ79SbsA85oeZg/8R6Gkruh6AFGFGkLD+I1vNXEvNuR1TTUarMCf94zwRGJObtC3ye0NDZgV8WmG8snU944nJ5m/CvO97/9yb8u19abNnAeHIJwXlEyX5UjodE8zvw6xLbzlLg90THh2jbmUAu0VvvWAj8Khq3Avgm0bEY+AQhwYrX0SeKlFMLIYG5jFwi2kCoNfBoG4hrN/Yj/Fb0ISQJ8XGnUFxnkksITksseyPgOHLHj/Py5vtLNPwp4NOJ4f2ARkLtW4f+DAEPkUs8vpBYr/sDc8kliIXW64bsmwWP9wXWw/XAZ4HNEsOHR9/VCUnG8Lz5vppYf/8GbJSIbTjwrQJluyH79MRo3I0d3q/KXEnzqXDCE72PN+TPJqb5bDTs6eh9OQlPnL1flzf8O9HwP3X2uxGq5uMfriEd2LDj77oIGFxg/MnkfvS2yxs3jdw/ip2KLL8/uX9AhWrJehKqCR14rNwdgFDb4cAfC+040TRTomn+K294/A96OgWqzIssKy6HaWVuV2Pzhl9E7gC9ZYH5Ppf4rp/NG3djYtx+BebdPTF+u3K+T4XWTRxXu/tbgXkbEjG/SIHaTXIHjGVA/w4suy8wK5r3wLxxY8j9GJ7SgWXG2/rJJaaJazOeKfR9ommujqb5RYl1/LkSn9GpMiec7lpNODDn15iMpRMJD3AQuWPOejWp0TpeXmjZ0bYV7ytHF1n+9oRkaDWwVZFt5yEK/3D+KRqf/8dlO3KJ9DcLzNeP8IPuwM1FysmBGwrMOzyxba2iwKkxwunyQnFtSkh2VwOfKVIee0XLX0LidF2ijPfsyPorsV73T3zPgwqMH0lIFjuzzRTdN6PxBY/3HfyMuIwvzhse73vXdGBZG7JPT6STCU/aHQ/eHD1/NTHsq3njSjKznuQuP78lb/RthA35sI5cVm5mfcxstJldRzgtAKFKrrncZST8yt2XFBh+M+Fccg/g6CLzPuDuLxYZ9zngY4QDwI/yR7r7WkI7JoD9zWzLMuM9KXqe7NHWVcCt0fMh8QAz25Fwagngu+5ercvdvxQ9X+fub+ePdPeHCKcOIdTgFPK4u/+1wLwzCOsIcldOlaOr1k1HXenuKwsM/wnhwPoxEuuwPdGyHo7e7ps3+kTCP+7Z7l7pKxrjbfKXRb4PwP9Gz8W+zwvRtlBR7v4K4YemH5W7tP6Y6Plv7v63Ap85n3BsK2QsIfmY7+53F5rA3V8l1Fr0iqYvZFKR/f+P0fNOBWLuQaiRuq7AZy4nty8cEx23C7m8wLyvEU4DQWijObfAfI8UietYQnvEv3qRKzzd/SlCm9JBhD85sfej562KxNpR8bHqKXd/tEAccwm1ah3Wzr5ZKX8qsvzOlFMl9ukOSzvh+T2hCvNYM+sXdfB3DCEbL7ZD5zsE2JpQLTk9OcLd3yRUpfYitKUp5eJEB4crCQexr0fjniKcW+6MaYUGuvs6wr98COeKC3myyPDkPDPdfWmRaR4jnF4o9Rmtoo4at43e3hH18rveg9BwGcI51dhe0fMSd3+6vc+qBDPrQ+4At94BJOEv0XOxMni2xLzxZfuDOhBaxddNJ00rNNDd3wf+XuyzzWxHM/uFmb1gZu+b2brEvhH3BZR/xWK8/u+vQNzJWHqRS6R/UmKbjH/ci3U2WmpfKieOQ8zsVjObZ2bL8zpE3SWarFJXcX46el4vCU94vMjwfeJYipVVVF7xj1ax8iq2TxTbH+Lt6PEooS8k3g/7E5or5IsvjChkcfRc7A/goiJxxeWxZzvlMTyaLlke8bZ8s5lNMrO9zKx3kc8vR1xG00tMU2pcZ/fNspnZYDO7yMyeiDrfXZNYfryP5S//geh5vJnda2bHRFdNF/uMSu3THdarUgvqDHdfamb3EbLwown/EDchNIp7t8zFxJnibVESke8W4OBoup+WWE4LucuA4wZzLxMK/TZ3X1NsxnaU6ucmHlesA8ZSNUrxPEWX7+4rzOxdwjnRop08JiQz9HKmT/ZAPTR6fq2M+SplMLmkvVQ5x7U0xb7TByXmXRE9d+RA1xXrpjM6vO2Z2fGE2sf4+64j19gewr/l/tEjqavW/2BCO434dXs2LjK8M7WzAJjZVcB/JgatJpz+iGsxBxPKK79MOmvz6LlUH0xvFhke78N9yK2TUgr2Iu/uxfaJYvtDu9s8uf0wOX3SohK1ynESVaxM4vH5ccXlsTHFt42kZHmcQ0jM9gHOjR4rzOxJ4A7CKZWPylhmLP7OxdYdlCi/Ddg3y2JmowlJaXK7+YBcw+U+hISyzfLdfbqZ/Tfw34R2bZ+Pljeb0Pzh1+6eTGQrtU93WLk1PPFG3pENptwNIXlaq6OnswYSGugBfKfALSKc0AgLYFcz+1SJxSU7HtzG3Ue7+7Hu/j8bkOy0+xXaGV/sn1JS30oEEkluDwPd3dp5NCSmb++7dLVKlkOl1GJMsfXWV9Tz+W8IB9TfExpmbuTug+J9g9B4sdD8XbX+k9vkLmVsk8XiKGdfWo+ZHU5IdtYS2g6MJLQ52CxRJnGNZqXKYEOWE5fX3eWUlbtPrEC8SaW2+WLJTFeKy2NymeVxY2uw4U/3foSzCFcRakX7ENpYXQ28aGbbUlkF1/0G7pvluoGQ7DxHuFBmE3f/mLsPjZZ/XLHlu/slhMbp5xGu1HufcKX02cBLZpZstlKpfbrDyk144tqWkufozKwvuYyt3BqaBwjX+x8cPZYQssJyfJmOZX8ntT9JxZWqXozLszP/PuN5tis2gZltBMRVi+V8xqLE69EdjCduPzO85FSVtYTwLwdKlAO503Sd/pffQV2xbjqjo9ve4YR/iS8RumCYUaAtVrFag3j9l1oPnfEuuWSlo9tkJcQH+evc/XvuPq9ALUQ5NSkdEa+TUsfbYuPifbjaZdXuNk/bUxPV2hc3qDw8mOruZ7r7boTat28Sjj0fJ5dklCP+zuXsl/k2ZN9sl5kNJ5xmWgt8wd3/7O75HZ+WXL67v+ruk9z9MEIecBDh1H0v4OpEO9rU9ulyE574fP8+JacKBRY3Rvt7qQlj0Uq7LZqvB/B7d19VZlxxAvNDQlVbsUd8C4t/jc4fVtOBhQaamRFa7UPIqDsqnmcHM9umyDQHkDtt2e5nRI0Z4wPEMaWmLeCp6Hmwme1Vcsq24oSlw1l8tJ3E5/QPKjHpZ6PnzpRzZ1R83XRSsW1vE3LtCZKfHSeGLxQ6PRxts5/NHx6J1//hHYyx5PqPjg9N0duObpOVEJdJweOZmW1HqPWppPiz9isxzf5FhsdtlUaZWUca2m+oeDva04rfbDnedloIl/NXQ1weB5ZqV1Iud1/qoVH++fFyOzB7XEYHlJim2PI2ZN+EXO1aseNs659CL367oXEllt/2w9zXuvs0Qt8+qwmnwRqjcRu6T3f6N6PchOcP0fMIMxtfYrrvRM+vUmbCE/kloZfcK4Gy7mVjZiPINby7zd3fK/YA7iHsZEMJ/epU02lW+A61/0b4x7OO0OFcRz1EqDbsTTjX3EZ0FcRF0dvHvcAVTEXcGD2fbmafLDaRBQPj9+4+m3CJIcCPOtC4L27h39mb2t0ZPZ9sZuv9OzKzzxH6LoHQ/001dNW66aizo4bd+b5N6IPk/SjWWHw39p2iA2i+/0foi6aQuMPBHc3smx2IsZz1f2P0fKyZlUpsMbOONC4vR1wmOxcZ/wMqfzovbqy5r5ntnT8y+jd+fJF5HyHXjmqyFb8aqtJldRfhWLYZcEqBz+pHbl+4q0TD5kq7g3Ds3wi4otSEyfIwsx7t/DmOm2x05LT1HdHz3ma2XtJjZh8n9BNVyIbsm9D+fhYvf6gVuKLZzHamyIU/RY4xsVXkanOSZXVj9NyZfbrzvxle/nXzcYdJHxCq9AYmxo0i3PMqvta/YI/EifFl905JkX54yPXU/EqZy7kjmv72vOHz6WSfJ+18Xvxd4y7od4qG9ybUTLVE40t1PHhyO59xLrn+Oi4g1xnVhnQ8uCm53p0XR7EOSIwfRti5nsuPj5CAxh2APULitgyEquDjgVvy5tkhmn41Jfq7oLyOB2fGn0moMTyW8joeLLruy10XFV437cZV4nMb8ra9KeQ6j+tH+FMS94B6ft68o8j1d/Jzch20fYzwY7WacPq5YB8YhC74nVzHg3GHdz2j9TwRODVvnvgGvY+TOKbkTdOb8E/dCfvNmST6tgK2INTiTssvs3LLsth0hB/vePv8GrnO0YYTetKO+29Zbxuh8/3wGOFUgEfb9uHkOvDci3Bao1THg19IrMdHgD0T8/ciXHo9CVhabNspEVvR75RY/yuicutox4NFy4l29sN24vpPcvvE7ST6NiMkQvsR/nS/lBi+KeGYcwEh2e0ZDe9BaHqxMF5eB9dt/Du6mFD7EXc8uC/h9hUFOx5kw/fNv0XjJsffJW98j8R3epSoryPCvncM4ZT1O0Viu43Q/udQEj2OR9vTbdE8y0nchoMN26fj2029C+zQofLvwIraPFFo8YF8CW1vObAOuKDEMuLpNijhIRwQXo2GX1Hmco4ntzMOSgyfT9cmPP9CLrl5j47dWqLgzp2Yrifr374gbtcS/6Ce3k58DQXGjSQcVD2xnHdpe8sIB04qUs4rEtMku8QvdkCanhj/brRO5gN7FVhPYwvM/xlyPzrO+reWmEnpW0sUXfflrosKr5t24yrxuQ2Jzyx1a4k/UvjWEj/JW8dLyCVID5LrmPLGAvP2JTSoTM6f/7kT8+bZkdw+sZpwlcp8Qr8pyem2IFymnX/8+SDv8y7uTFkWm47QSPVJ2q7H5O0FLiq2jbBhPS0PJ9fjfP5+9Da5WxDMKTL/v9P2WPMR4Qcr2d2/F9t2SsRV9DsREuqHEsvPvxXDCtq5tURn98P2lkHo7Tx5O5CWaPtJ3l7i1cT0m+ZtV6sIx6Zk+c0Dtu3ges2/tUT+LUNK3VpiQ/bNf8/bFhYQ9rMfJ6Y5Oq883k9sQwsIZyXWi41wLEnul0vJ/ebF+0yhW1J0dp/uTa4Ty3VRuc2PHiXXR9n98Lj7O4TziycS/jUuIjSignA+9jeEGyJeVu4yN8CBhJ0Tyj8dNIWw8vpSvNqwKzxB+Id1O7mNZw7hEr6xvn7DsLJ5OE96EqFDq4cICdUAwqWbtxJ6Fr26E8udS+gL5HRCtr+E8E9iDeH2DT8nrIPfFZj3NsK9uH5B+McCYaN8mdAh2Vfz5yH8g7iakMQOIDR83I7w76uceJ8hNH6bHH1m7yjWJsK/nz3dfXHxJVReV62bDsbwB0LbpimEA9kaQvL3n4R71Kx39aG7f4fw7/zvhO21F+F+N98m3OKg6BWL7r7S3b9CuHLyT4RjRH/CD+1ThH/Lv8mbZzbhH9uDhGr1LQnrftu86RYTtrl/JfSPsphQnka41cBvgSMIp5gqxkM7sXGEGpFXCNvyGkInb5/3cHVKxXnobG83wtVBrxES6PcI5bc7uYtC3isy/w2EWoGfEvoUW0O4bcy7hH36v8gdQysV83JCbdQ3CDV2ywlJ0ALCvr+zu99Tyc/sQGyXEvpLupaQcBhh23yLcOHMaYTjdOx9Qg3MTwmn6psJXaa0EPoougDY1d2Tl9qXE8dbhFvu/IRQLj0J2/1vCet7Xol5N2TfvIFQM/9MNN0wwn62eWKauwntgB4mJB29oxh/TPg9KPZdJwDfJezDrxD+JPSMvssNhLyg0G9Fp/ZpD22ADib8/rxBqOWPfzNKttGNqzmlwqJL4iHcx2h+mrHki85Nxy38t452QhGpE2Z2CaHW4iZ3PznlcETqQto9LUs64ssi1xJqb0SkTpjZYHK9wD9caloRyVHC081EtTtnRW9nevH7mIhISsxsTzP7uZk1Rn02YWa9zOyzhFNSWxHaLPyhxGJEJCHVW0tIdZnZDYTzpfEl4x3pNEtEqmcT4D+iB2a2lNDmJL4EeAnwFXdfUXh2EcmnGp7uZRCh0fQ/gK+7+/+kHI+IFPY8oY3OdMLlwv0I7e5mEfor28mL3P1bRApTo+UM23zzzb2hoSHtMKQCXnxzGclddedtBhafuBPe+XAlby0rXlkQf96adc7SllUM6t+HXj3Svn2aSOXNmDHjHXfvqhv6Sop0SivDGhoaaGpqan9CqXkNE3K3l9ukDzR9/8iKf8b+l09l4bLCTbp+f/aBjBgyoOA4kSwxswVpxyBdQwmPSB34zriRTJ46l7PGjeSMcaMquuwlLau46Yn5HNM4jDtnvM4b761f03PwldPbvL/wiB35xgGlerIXEaktOqWVYY2Nja4aHiklv1bnq3ttx81Ptf8Ht4fBK5dXvpZJJG1mNsPdG9OOQypPNTwiVbSkZRW7XZLrOqVh0EZMO/fg1OJJJjtnHjySk/bZns0H9Gby1LnsO2Iwf51XuJumsZ9QEwcRqS9KeESq6IRf/63N+/lL072qeNjAvixctpJhA/ty1iHhVNkZ40a1njYb+8NHWmM8YMRgFixdwYGjhvDtcZ9ILWYRkc5QwiNSRXMWL2/zvmFQWbcL6zKPnzeu5Pg0a59ERCpJ/fDUGDMbZmaPmtnLZjbLzM6Mhg82s4fN7J/R86C0Y5WOO2LMFq2vvzNupBIKEZEqUaPlGmNmWwFbuftzZrYJMAP4InAysMTdJ5nZBGCQu59ballqtFw/zrp1BnfPfJujd9mSySfsnnY4It2WGi1nl2p4aoy7v+Xuz0WvPwBeBrYBxgM3RZPdREiCJCPunvl2m2dp66qpc9h+whSumjon7VDa1TBhSuujFpYjIoESnhpmZg3Ap4GngaHu/haEpAjYosg8p5hZk5k1NTc3VytU2UBH77Jl6/PI6EdupH7oWk2eOhePnkVEOkMJT40yswGEOyF/293fL3c+d7/W3RvdvXHIEF06XC8mn7A78ycdyeQTdmdNNGxNyTm6l7PGjcSiZxGRztBVWjXIzHoTkp1b3P2uaPAiM9vK3d+K2vksTi9C6Uq9CMmOds6c5KXytW7+pMp0yFip5YhIoGNqjTEzA34LvOzuP0mMuhc4CZgUPd+TQnhSBXP1QyciUnFKeGrPvsCJwD/M7Plo2PmEROd2M/s68BpwXErxiYiI1B0lPDXG3f8KWJHR6rRFRESkE9RoWURERDJPCY+IiIhknhIeERERyTwlPCIiIpJ5SnhEREQk85TwiIiISOYp4REREZHMU8IjIiIimaeER0RERDJPCY+IiIhknhIeERERyTwlPCIiIpJ5SnhEREQk85TwiIiISOYp4REREZHMU8IjIiIimaeEp8aY2fVmttjMXkwMm2hmb5jZ89HjiDRjFNkQE+58noYJU5hw5/NphyIi3YgSntpzI3BYgeGT3X3X6HF/lWMSafXD+1+iYcIUfnj/S52a/7amN1qfGyZM4ePnTeHWpxdUMkQRkfUo4akx7v4YsCTtOESKueaxV9s8d9TQTfq2eb/O4by7X+TQKx/d4Nik+nY8fwoNE6aw4/lT0g5FpKReaQcgZfsPM/sq0ASc7e5LC01kZqcApwAMHz68iuFJFp3+u2e5f9bi1veNwz7GaQdszzWPvcppB2zfqWX+7yl7cfCV09cbPqd5eafjlPSsWNf2WaRWqYanPlwDjAB2Bd4Criw2obtf6+6N7t44ZMiQasUnGZVMdgCaFr7PuUeMZv6kIzn3iNGdWuaIIQN47qJDGDGkX5vho/LeS2Fn3TqDhgmhVuXTEx9IOxw26tH2WaRWqYanDrj7ovi1mf0GuC/FcKQbOWLMFm2Snq036VOR5Q7u34dHzj6oIsvqbu6e+Xbr66U1UK0y+wdHph2CSFmU8NQBM9vK3d+K3h4NvFhqepFKufrEPQBY0rKKO5oWclzjsJQjkqN32bI16RmkahWRspm7px2DJJjZrcBYYHNgEXBx9H5XwIH5wDcTCVBRjY2N3tTU1FWhiohkjpnNcPfGtOOQylMNT41x9xMKDP5t1QMRERHJENWHioiISOYp4REREZHMU8IjIiIimaeER0RERDJPCY+IiIhknhIeERERyTwlPCIiIpJ5SnhEREQk85TwiIiISOapp2URkTqyz2UP8+YHqwC48eQ9GLvjFilHJFIfVMMjIlJH4mQH4PT/1b3yRMqlhEdEpI5svUmf1tdr11qKkYjUFyU8IpKqJS2r2OP7f6ZhwhSOnDwt7XBq3hMXHMKFR+xID4NzDv1E2uGI1A214RGpA3t8/880L1/TZtjGveC+Mw9kxJABKUVVGXc0LWz9brMWtTD6wim8dOmRKUdVXeOvms7MNz9k6IA+PHDWgQzu36fk9A+9tIh1DpPun82l989m4EbGo+eMY5/LHmbFOtioB8z+QfcqQ5H2KOERqQP5yQ7AR2vgsCunswY4a9xIzhg3qvqBVcBxjcO4/IHZre8LfNXMm/nmhwAs+nAVdzQt5K6m15jTvLzNNLuGGIVeAAAY/0lEQVQNG0i/vr353vgxfLhyLQBxUS1b4Zxx699ZsS68j59FJEcJj0gdGNKvV8GkZ3X0PHnq3LpNeAb378NpB2zPNY+92qH5xlw0hZaoAI5v3IZHX17MopbVDO3fm6cv+lwXRNp1dtl6QGsNT34CGHtu4TIADr5yOn0KNEb469x3Wl+rZY/I+pTw1Bgzux44Cljs7jtFwwYDvwcagPnAl919aVoxSvU9+9+HMuHO57mt6Y02w4cN7Mvry1Zy1riRKUVWGeceMZov7TGcS+97iQuPGl3WPHGyA7Qpl0XJEXXinjMObPO+h8E6Lz79qgI1OPuN3JzRW23Cxn16cdI+DZUNUCQDzL3EXiVVZ2YHAB8CNycSnh8BS9x9kplNAAa5+7ntLauxsdGbmnTZqmRTlmp48k2bvZizbn+eQz65Bd8cO5KDr5ze7jzPXXRIu21/pH1mNsPdG9OOQypPCU8NMrMG4L5EwjMHGOvub5nZVsA0d2/3/IUSHpFsuGrqHH4yde56w0cP7c9Li1rYs2EQvz91nxQiyx4lPNmlhKcGFUh43nP3TRPjl7r7oCLzngKcAjB8+PDdFyxY0PUBi0hVTJu9mK/f9CxrHQ4YMZif/svu3NG0kOMah6l2p0KU8GSXEp4atCEJT5JqeEREOkYJT3ap48H6sCg6lUX0vDjleEREROqKEp76cC9wUvT6JOCeFGMRERGpO0p4aoyZ3Qo8CYwys9fN7OvAJOAQM/sncEj0XkRkPbc+vYCR59/Pqb9rYknLqvZnEOkm1IYnw9SGR6T7aZgwpfV1D+CVSbrFREeoDU92qYZHRCSj1gF/ev6NdqcT6Q6U8IiIZMh38nrdPufOF1KKRKS2KOEREcmQM8aNYpetB7S+P3TM0BSjEakdupeWiEjG3HPGgSxpWdXaKaGIKOEREcmkwf378M0DR6QdRpc74Vd/48n577F3w6bceuq+aYcjNUyntEREpG49Of+91udDr3w05WiklinhERHJqKumzqFhwhQaJkzhrFtnpB1Ol5vTvJznFixNOwypUUp4REQyKnmH9btnvp1iJF2nf++278+5c2Y6gUjNUxseEZFuoGfaAeQZfeEUlq+Bfr3gpUs73znirEvCvM8tWMo5d87kii/tUqkQJWNUwyMiklFHjNmi9fXaFOMoZPmats8barftBvHI2WPZbbtBlVmgZI4SHhGRjLr0mNqt7ejXq+2zSFfTpiYiklF3NC1MO4SiNuQ0lkhnqIZHRCSjjmscxiZ9wuthA/umG4xIylTDIyKSUYP79+Ef31dNigiohkdERES6AdXw1BEzmw98QLjgYo27N6YbkYiISH1QwlN/DnL3d9IOQkREpJ7olJaIiIhknhKe+uLAQ2Y2w8xOKTSBmZ1iZk1m1tTc3Fzl8ERERGqTTmnVl33d/U0z2wJ42Mxmu/tjyQnc/VrgWoDGxkZPI0iR6x6bx6X3z2593xN49qJDGNy/T3pBiUi3phqeOuLub0bPi4G7gc+kG5FIYclkB0Ir+/FXPVZ4YhGRKlANT50ws/5AD3f/IHr9OeD7KYclUtDGvXvw0ep1bYYtXLayzfs9vv9nmpevoafBQ985kCsffJn7Zy3miDFbcPWJe1QzXBHpBlTDUz+GAn81s5nAM8AUd38w5ZhECrrmX3dncP/enHbA9q3D9m7YtM00zdFdI9c6XHrfS9w/azEA989azLTZi6sXrIh0C+auZh5Z1djY6E1NTWmHIVJQfg3P53/2GMvXhOPRxr178KNjP8V/3TmTlWtyx6jjG7dh0pd2TStk6QbMbIb6OMsmJTwZpoRH6slzC5Zy+i0zWNKyilVrnT49YNW69aebP0m3SpCuo4Qnu3RKS7qV/S+fSsOEKewy8UGWtKxKOxxJ2G27QTx1/jhOGzsSKJzsiIh0lhIe6VbihrPLVqzljqaFKUcjhZy0TwMHjRqSdhgikjFKeKTbOq5xWNohSAGD+/fhyi/vSl9bf9zPj1f7HRHpHF2WLpm1y8X3s2yl07cHPHlB6PTurtP24Zw7Z3LFl3apaid4S1pW8e/XP83MN95nQN+e3Py1Pdltu0FV+/x6M7h/H+ZcfiTzmj/kv//4ImO2GcipB45Qx4Ui0mlqtJxh3b3RcsOEKa2vzzt8R7554Igu/8x5zR9y8JXTW9/fddo+NGzen70ufZhViV1txJD+PHL22C6PR0Q6Ro2Ws0s1PJJZA/taaw1PtU5fXXrfS23eH3PNE4wY0q9NsgNwxZd2qUo8IiISKOGRzJr5vSOAcDrpjqaFHNc4rMtPiVx41GgenTO9zbB5zcvbvN9s4546nSUiUmVKeCTzbnriVX72yFx+NX0e3z10FBfe/SJrgaH9e/PeR6tZmbj8+bQDtufcI0Z3+rNGDBnAjSfvwb/f+CzJSp2+PWDMtptyxXG7MGLIgE4vXyTLTv/ds7q9iHQZXaUl3UC43Gfp8tWcFyU7AIta2iY7ANc89uoGf9rYHbfg1UlHctdp+7QOW7kOBm7cuyaTnSUtq9j54vtpmDCF/S+fmnY40o0lby8iUmlKeCTzTtqnoc37ntHz0P696Zu3ByTv/bSh8k9bXXhU52uOusr4q6az2yUP88HKUB+Vf4NPkWo6YswWbZ5FKkmntCTzBvfvw5ih/Zm1qIUxQ/sz5ayxbcZ3ZRufnsDa6LkWa3dmvvlhm/fDBvZNKRIRdBpLupQuS8+w7n5ZuhQXJ3n3PreQWYtaADh57+FMHL9zypGJpEuXpWeXanhEuqETf/MEs95uYcyW/XUzThHpFtSGR6QbmvV2S5tnEZGsU8Ij0s3s8f0/t74+epctU4xERKR6lPDUETM7zMzmmNlcM5uQdjxSn5qXr2l9PfmE3VOMRESkepTw1Akz6wn8EjgcGA2cYGa1d52z1Lwh/Xq1eRYR6Q50xKsfnwHmuvsrAGZ2GzAeeKnkXCKEm5qeeN3TvLlsha7GEpFuSTU89WMbYGHi/evRsDbM7BQzazKzpubm5qoFJ7VrScsqDr5yOm8uWwHAjU++lnJEIiLVp4SnfliBYet1ouTu17p7o7s3DhkypAphSS27auocdrvk4TbDthm4UUrRiIikRwlP/XgdGJZ4vy3wZkqxSJ2YPHVum/e9DW7+xp4pRSMikh4lPPXjWWAHM9vezPoAxwP3phyT1Lizxo0EoGGzfjxy9oH88/Ija/IWFyIiXU2NluuEu68xs/8A/ky4NdP17j4r5bCkxp0xbhRnjBuVdhgiIqlTwlNH3P1+4P604xAREak3OqUlIiIimaeER0RERDJPCY+IiIhkntrwiEgqdrn4fpatdAb2NWZ+74i0w6m4PS95iEUtqxnavzdPX/S5tMMR6fZUwyMiqVi20ts8Z8lzC5ayqGU1QOuziKRLCY+IpO703z2bdggVM2LCFI655om0wxCRPEp4RCQVR++yZevr+2ctTjGSylqb9/7yo3cqOu2SllX8evo8lrSs6tqgREQJj4ikY/IJu/OJIf0AWp+zoGfi9bBBG3PoTlsVnfb8u17g8gdms9slD3PdY/O6PjiRbkwJj4ik5rZT9+W8w3fktlP3TTuUipk36Ui2jm7QunDpR+x2ycNcNXVOwWmnvpyr2frBA7OrEp9Id6WrtKTbGX/VdGa++SEADYM2Ytq5B6ccUfcz6rwprHToazDn8iPTDqeizrp1Bm8uW9Fm2OSpcwve4uOS8WM47+4XAVjn0DBhCgD7j9iMn/3Lbgzu36frAxbpJlTDI91OnOwAzF+6osSU0lXiC7NWOvzp+TfSDabC7p759nrD4pu45vvMxzcrOPzxee9y0xPzKxmWSLenhEe6nV22bnu38FufXpBSJN2XJV5/9w8vpBZHV0g2xgbYbOOeRW/gOu7K6SWWlL3L9UXSpIRHup17zjiwzfuL7tFN56ttxkWHMH6Xrdiodw9+dOynWoc3TJjS+qhXk0/YnZP3Hg7AyXsPZ8bFhxWdtlRK84Vdt6lwZCLdm9rwSLfUMGij1tNZl4wfk3I03c/g/n342Qm7pR1Gl5k4fmcmjt+55DTzmj9cb1hPcpe1X3rfS1z55V3VjkekQpTwSLekhsqStpN/+8x6w07cezhnjBvFZ3/8Fx6d08xulzzMASMGc/P/2zuFCEWyRae0RKRmzJ90ZOsj6xa+99F6w2588jUG9+/Dex/lui98bN4SDr5yGs8tWFrN8EQyRwlPHTCziWb2hpk9Hz2yd6dFkW7mtAO2Lzj84Cv+wgEjBrcZNq+5hWOueUKdE4psAHPXlQC1zswmAh+6+487Ml9jY6M3NTV1TVAiUjEjzpvC2sSheP6kI3luwVLOuXMm85pbWof3MHglY/0W1Rozm+HujWnHIZWnGh4RkZRd+sXc/bZGbLYxALttN4hHzh7LuFGbt447//Adqx6bSFaohqcORDU8JwPvA03A2e5e8IS+mZ0CnAIwfPjw3RcsUB8zIiLlUg1PdinhqRFmNhXYssCoC4CngHcI3XZcAmzl7l9rb5k6pSUi0jFKeLJLl6XXCHcfV850ZvYb4L4uDkdkg3zjhqeZOucdRmy+MXectp/6khGR1KkNTx0ws60Sb48GXkwrFpFyTJ3zDgDz3gl3C//EBVMyd88sEakvSnjqw4/M7B9m9gJwEHBW2gGJlDKwr7V5v2pt9u6ZJSL1Rae06oC7n5h2DCIdsWxlrm1gD6BXT9rcM0tEpNqU8IhIxQ3cqCfLVqxl4EY9mTmx+M0zRUSqRae0RKTi9h6xeZtnEZG0KeERkYpZ0rKKH0x5mQdnLQLgwVmLGPvDR1jSsirlyESku9MpLRHpkN2/9yDvRje3/Pnxu/L4P9/h3hfeYMVqZ3C/nixZvrbN9POXruCYXz7OtO/qDvUikh4lPCLSIe8m7uT93T+8wEer17W+z092YvOXrChr2UtaVjH+qsdYuGzleuOO3mVLJp+wewejFREJdEpLRDpks417tr7+0bGf4su7b8tGva3EHKx39+9i7mhaWDDZAbh75tvlBykikke3lsgw3VpCqun03z3L/bMWA9AwaCPmL13B8Y3bMOlLu5a9DNXwZN/+l09l4bKVDBvYl8fPK6uD+arSrSWySwlPhinhkXq2pGUVdzQt5LjGYbo1RYY0TJjS+vqgUUO48KjRnHv732la+D6Nwz7Gnd/aP8XolPBkmdrwiEhNGty/D988cETaYUiF9e0JK6OmXo/OaWb6nOnErcCaFr6fWlySfWrDIyIiVXP/tw/koFFDWt+vA7Yd2BeAxmEfSykq6Q5UwyMiIlUzYsgAdhw6gEfnNLcOu/eMA3TaUrqcanhERKSqrnns1TbvlexINSjhERERkcxTwiMiIlV1xJgtWl8P6aeWFVId2tJERKSqrj5xj7RDkG5INTw1xMyOM7NZZrbOzBrzxp1nZnPNbI6ZHZpWjCIiIvVINTy15UXgGODXyYFmNho4HhgDbA1MNbNPuHvhGxeJiIhIG6rhqSHu/rK7zykwajxwm7uvdPdXgbnAZ6obnYiISP1SwlMftgEWJt6/Hg1bj5mdYmZNZtbU3NxcaBIREZFuR6e0qszMpgJbFhh1gbvfU2y2AsMK3gTN3a8FroVwL61OBSkiIpIxSniqzN07c3vg14FhiffbAm9WJiIREZHs0ymt+nAvcLyZ9TWz7YEdgGdSjklERKRuKOGpIWZ2tJm9DuwNTDGzPwO4+yzgduAl4EHgW7pCS0REpHw6pVVD3P1u4O4i4y4DLqtuRCIiItmgGh4RERHJPNXwiIgUsOP5U1ixLrx+5OwDGTFkQLoBicgGUQ2PiEgBcbID8PUbnk0vEBGpCCU8IiLtmL9kedohiMgGUsIjIlJAw6CNWl9/Z9zIFCMRkUpQGx4RkQKmnXtw2iGISAWphkdEREQyTwmPiIiIZJ4SHhEREck8JTwiIiKSeUp4REREJPOU8IiIiEjmKeERERGRzFPCIyIiIpmnhEdEREQyTz0ti4gAJ/zqbzw5/z32btiUW0/dN+1wRKTCVMNTQ8zsODObZWbrzKwxMbzBzD4ys+ejx6/SjFMki56c/16bZxHJFtXw1JYXgWOAXxcYN8/dd61yPCLdwugLp7S+3mXrASlGIiJdRQlPDXH3lwHMLO1QRLqNnf97CsvX5N4P3mTj9IIRkS6jU1r1Y3sz+7uZTTez/YtNZGanmFmTmTU1NzdXMz6RuvTBqrbvLzxqdDqBiEiXUg1PlZnZVGDLAqMucPd7isz2FjDc3d81s92BP5rZGHd/P39Cd78WuBagsbHRKxW3SFq+ccPTTJ3zTuv7nx+/K5/fdZuKLX+TPiHpMWDq2QcyYohOaYlkkRKeKnP3cZ2YZyWwMno9w8zmAZ8AmiocnkjNSSY7AN/9wwsVTXj+8f0jK7YsEaldOqVVB8xsiJn1jF5/HNgBeCXdqESqo1feUepHx34qnUBEpK4p4akhZna0mb0O7A1MMbM/R6MOAF4ws5nAncCp7r4krThFqum6r+7BoH69+fLu2/LcRYdUtHZHRLoPc1czj6xqbGz0piad9RLpiIYJuUvU50/S6a7uxsxmuHtj+1NKvVENj4jUpWmzF7PbJQ8xbfbitEMRkTqghEdE6s685g/5xs3PsqRlNWf+/u9phyMidUBXaYlI3bn4nhdZsy68/uijNSxpWcXg/n0qsmydxhLJJtXwiEjNOPiKv9AwYQoHX/GXktON3mpg6+tVwE1PvNrFkYlIvVPCIyI1Y967H7V5LubUsSPYrF/v1vc/e2Qu85o/7NLYRKS+KeERkZoxYrON2zwXM7h/Hx4+e2ybYZfe91JXhSUiGaA2PCJSMx4557NlT5vfZkf3wBKRUpTwiEjdUgNjESmXTmmJiIhI5inhERERkcxTwiMiIiKZp4RHREREMk8Jj4iIiGSeEh4RERHJPCU8IiIiknlKeERERCTzlPDUEDO7wsxmm9kLZna3mW2aGHeemc01szlmdmiacYqIiNQbJTy15WFgJ3f/FPB/wHkAZjYaOB4YAxwGXG1mPVOLUkREpM4o4akh7v6Qu6+J3j4FbBu9Hg/c5u4r3f1VYC7wmTRiFBERqUdKeGrX14AHotfbAAsT416Phq3HzE4xsyYza2pubu7iEEVEROqDbh5aZWY2FdiywKgL3P2eaJoLgDXALfFsBab3Qst392uBawEaGxsLTiMiItLdKOGpMncfV2q8mZ0EHAUc7O5xwvI6MCwx2bbAm10ToYiISPbolFYNMbPDgHOBL7j78sSoe4HjzayvmW0P7AA8k0aMIiIi9Ug1PLXlF0Bf4GEzA3jK3U9191lmdjvwEuFU17fcfW2KcYqIiNQVJTw1xN1Hlhh3GXBZFcMRERHJDJ3SEhERkcxTwiMiIiKZp4RHREREMk8Jj4iIiGSe5bp6kawxs2ZgQcphbA68k3IMtUJlkaOyyFFZ5NRCWWzn7kNSjkG6gBIe6VJm1uTujWnHUQtUFjkqixyVRY7KQrqSTmmJiIhI5inhERERkcxTwiNd7dq0A6ghKosclUWOyiJHZSFdRm14REREJPNUwyMiIiKZp4RHREREMk8Jj3QJM7vCzGab2QtmdreZbZoYd56ZzTWzOWZ2aJpxVoOZHWdms8xsnZk15o3rVmUBYGaHRd93rplNSDueajKz681ssZm9mBg22MweNrN/Rs+D0oyxGsxsmJk9amYvR/vGmdHwblcWUj1KeKSrPAzs5O6fAv4POA/AzEYDxwNjgMOAq82sZ2pRVseLwDHAY8mB3bEsou/3S+BwYDRwQlQO3cWNhHWdNAF4xN13AB6J3mfdGuBsd/8ksBfwrWg76I5lIVWihEe6hLs/5O5rordPAdtGr8cDt7n7Snd/FZgLfCaNGKvF3V929zkFRnW7siB8v7nu/oq7rwJuI5RDt+DujwFL8gaPB26KXt8EfLGqQaXA3d9y9+ei1x8ALwPb0A3LQqpHCY9Uw9eAB6LX2wALE+Nej4Z1R92xLLrjd27PUHd/C0IiAGyRcjxVZWYNwKeBp+nmZSFdq1faAUj9MrOpwJYFRl3g7vdE01xAqL6+JZ6twPR13zdCOWVRaLYCw+q+LNrRHb+zFGFmA4A/AN929/fNCm0eIpWhhEc6zd3HlRpvZicBRwEHe67Dp9eBYYnJtgXe7JoIq6e9sigik2XRju74nduzyMy2cve3zGwrYHHaAVWDmfUmJDu3uPtd0eBuWRZSHTqlJV3CzA4DzgW+4O7LE6PuBY43s75mtj2wA/BMGjHWgO5YFs8CO5jZ9mbWh9Bo+96UY0rbvcBJ0euTgGI1gplhoSrnt8DL7v6TxKhuVxZSPeppWbqEmc0F+gLvRoOecvdTo3EXENr1rCFUZT9QeCnZYGZHAz8HhgDvAc+7+6HRuG5VFgBmdgTwU6AncL27X5ZySFVjZrcCY4HNgUXAxcAfgduB4cBrwHHunt+wOVPMbD/gceAfwLpo8PmEdjzdqiykepTwiIiISObplJaIiIhknhIeERERyTwlPCIiIpJ5SnhEREQk85TwiIiISOYp4REREZHMU8IjIiIimff/AdiUkVxcdePhAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "#plt.scatter(*embedding.T, s = 1)\n",
    "plt.scatter(embedding[:, 0], embedding[:, 1],  s = 1) #c=[sns.color_palette()[x] for x in Y])]\n",
    "#plt.gca().set_aspect('equal', 'datalim')\n",
    "axes = plt.gca()\n",
    "axes.set_xlim([-25,25])\n",
    "axes.set_ylim([-22,22])\n",
    "plt.title('UMAP projection of bacterial genomes dataset', fontsize=24);\n",
    "\n",
    "'''\n",
    "Where a and b are assumed to be 2 points in the UMAP space\n",
    "\n",
    "'''\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then, we cluster the data using HDBScan, using it to fit our embedding. The clusterer we get has labels for all 3867 genomes, ranging from cluster 0 to cluster 39 (with cluster -1 labeling data that was unable to be clustered). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import hdbscan\n",
    "\n",
    "clusterer = hdbscan.HDBSCAN(min_cluster_size=10, min_samples = 40).fit(embedding)\n",
    "print(clusterer.labels_.max())\n",
    "print(clusterer.labels_)\n",
    "print(clusterer.labels_.min())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From the cluster labels, we want to plot our reduced data while associating each point with its respective cluster, and evaluate its closeness to each data point sharing that cluster. We thus import umap.plot, a package using the interactive Bokeh plotting tool, and plot each genome in the embedding, with a different color according to the clusterer’s label of that genome. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bk-root\">\n",
       "        <a href=\"https://bokeh.org\" target=\"_blank\" class=\"bk-logo bk-logo-small bk-logo-notebook\"></a>\n",
       "        <span id=\"1467\">Loading BokehJS ...</span>\n",
       "    </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  var force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "  var JS_MIME_TYPE = 'application/javascript';\n",
       "  var HTML_MIME_TYPE = 'text/html';\n",
       "  var EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  var CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    var script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    var cell = handle.cell;\n",
       "\n",
       "    var id = cell.output_area._bokeh_element_id;\n",
       "    var server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      var cmd = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            var id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      var cmd = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    var output_area = handle.output_area;\n",
       "    var output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!output.data.hasOwnProperty(EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    var toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      var bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      var script_attrs = bk_div.children[0].attributes;\n",
       "      for (var i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      var toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      var props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    var events = require('base/js/events');\n",
       "    var OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  \n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  var NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    var el = document.getElementById(\"1467\");\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error() {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (var i = 0; i < css_urls.length; i++) {\n",
       "      var url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error;\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    const hashes = {\"https://cdn.bokeh.org/bokeh/release/bokeh-2.0.1.min.js\": \"JpP8FXbgAZLkfur7LiK3j9AGBhHNIvF742meBJrjO2ShJDhCG2I1uVvW+0DUtrmc\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.0.1.min.js\": \"xZlADit0Q04ISQEdKg2k3L4W9AwQBAuDs9nJL9fM/WwzL1tEU9VPNezOFX0nLEAz\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.0.1.min.js\": \"4BuPRZkdMKSnj3zoxiNrQ86XgNw0rYmBOxe7nshquXwwcauupgBF2DHLVG1WuZlV\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.0.1.min.js\": \"Dv1SQ87hmDqK6S5OhBf0bCuwAEvL5QYL0PuR/F1SPVhCS/r/abjkbpKDYL2zeM19\"};\n",
       "\n",
       "    for (var i = 0; i < js_urls.length; i++) {\n",
       "      var url = js_urls[i];\n",
       "      var element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error;\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      if (url in hashes) {\n",
       "        element.crossOrigin = \"anonymous\";\n",
       "        element.integrity = \"sha384-\" + hashes[url];\n",
       "      }\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };var element = document.getElementById(\"1467\");\n",
       "  if (element == null) {\n",
       "    console.error(\"Bokeh: ERROR: autoload.js configured with elementid '1467' but no matching script tag was found. \")\n",
       "    return false;\n",
       "  }\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  \n",
       "  var js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.0.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.0.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.0.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.0.1.min.js\", \"https://unpkg.com/@holoviz/panel@^0.9.5/dist/panel.min.js\"];\n",
       "  var css_urls = [];\n",
       "  \n",
       "\n",
       "  var inline_js = [\n",
       "    function(Bokeh) {\n",
       "      inject_raw_css(\".json-formatter-row {\\n  font-family: monospace;\\n}\\n.json-formatter-row,\\n.json-formatter-row a,\\n.json-formatter-row a:hover {\\n  color: black;\\n  text-decoration: none;\\n}\\n.json-formatter-row .json-formatter-row {\\n  margin-left: 1rem;\\n}\\n.json-formatter-row .json-formatter-children.json-formatter-empty {\\n  opacity: 0.5;\\n  margin-left: 1rem;\\n}\\n.json-formatter-row .json-formatter-children.json-formatter-empty:after {\\n  display: none;\\n}\\n.json-formatter-row .json-formatter-children.json-formatter-empty.json-formatter-object:after {\\n  content: \\\"No properties\\\";\\n}\\n.json-formatter-row .json-formatter-children.json-formatter-empty.json-formatter-array:after {\\n  content: \\\"[]\\\";\\n}\\n.json-formatter-row .json-formatter-string,\\n.json-formatter-row .json-formatter-stringifiable {\\n  color: green;\\n  white-space: pre;\\n  word-wrap: break-word;\\n}\\n.json-formatter-row .json-formatter-number {\\n  color: blue;\\n}\\n.json-formatter-row .json-formatter-boolean {\\n  color: red;\\n}\\n.json-formatter-row .json-formatter-null {\\n  color: #855A00;\\n}\\n.json-formatter-row .json-formatter-undefined {\\n  color: #ca0b69;\\n}\\n.json-formatter-row .json-formatter-function {\\n  color: #FF20ED;\\n}\\n.json-formatter-row .json-formatter-date {\\n  background-color: rgba(0, 0, 0, 0.05);\\n}\\n.json-formatter-row .json-formatter-url {\\n  text-decoration: underline;\\n  color: blue;\\n  cursor: pointer;\\n}\\n.json-formatter-row .json-formatter-bracket {\\n  color: blue;\\n}\\n.json-formatter-row .json-formatter-key {\\n  color: #00008B;\\n  padding-right: 0.2rem;\\n}\\n.json-formatter-row .json-formatter-toggler-link {\\n  cursor: pointer;\\n}\\n.json-formatter-row .json-formatter-toggler {\\n  line-height: 1.2rem;\\n  font-size: 0.7rem;\\n  vertical-align: middle;\\n  opacity: 0.6;\\n  cursor: pointer;\\n  padding-right: 0.2rem;\\n}\\n.json-formatter-row .json-formatter-toggler:after {\\n  display: inline-block;\\n  transition: transform 100ms ease-in;\\n  content: \\\"\\\\25BA\\\";\\n}\\n.json-formatter-row > a > .json-formatter-preview-text {\\n  opacity: 0;\\n  transition: opacity 0.15s ease-in;\\n  font-style: italic;\\n}\\n.json-formatter-row:hover > a > .json-formatter-preview-text {\\n  opacity: 0.6;\\n}\\n.json-formatter-row.json-formatter-open > .json-formatter-toggler-link .json-formatter-toggler:after {\\n  transform: rotate(90deg);\\n}\\n.json-formatter-row.json-formatter-open > .json-formatter-children:after {\\n  display: inline-block;\\n}\\n.json-formatter-row.json-formatter-open > a > .json-formatter-preview-text {\\n  display: none;\\n}\\n.json-formatter-row.json-formatter-open.json-formatter-empty:after {\\n  display: block;\\n}\\n.json-formatter-dark.json-formatter-row {\\n  font-family: monospace;\\n}\\n.json-formatter-dark.json-formatter-row,\\n.json-formatter-dark.json-formatter-row a,\\n.json-formatter-dark.json-formatter-row a:hover {\\n  color: white;\\n  text-decoration: none;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-row {\\n  margin-left: 1rem;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-children.json-formatter-empty {\\n  opacity: 0.5;\\n  margin-left: 1rem;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-children.json-formatter-empty:after {\\n  display: none;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-children.json-formatter-empty.json-formatter-object:after {\\n  content: \\\"No properties\\\";\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-children.json-formatter-empty.json-formatter-array:after {\\n  content: \\\"[]\\\";\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-string,\\n.json-formatter-dark.json-formatter-row .json-formatter-stringifiable {\\n  color: #31F031;\\n  white-space: pre;\\n  word-wrap: break-word;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-number {\\n  color: #66C2FF;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-boolean {\\n  color: #EC4242;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-null {\\n  color: #EEC97D;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-undefined {\\n  color: #ef8fbe;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-function {\\n  color: #FD48CB;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-date {\\n  background-color: rgba(255, 255, 255, 0.05);\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-url {\\n  text-decoration: underline;\\n  color: #027BFF;\\n  cursor: pointer;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-bracket {\\n  color: #9494FF;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-key {\\n  color: #23A0DB;\\n  padding-right: 0.2rem;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-toggler-link {\\n  cursor: pointer;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-toggler {\\n  line-height: 1.2rem;\\n  font-size: 0.7rem;\\n  vertical-align: middle;\\n  opacity: 0.6;\\n  cursor: pointer;\\n  padding-right: 0.2rem;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-toggler:after {\\n  display: inline-block;\\n  transition: transform 100ms ease-in;\\n  content: \\\"\\\\25BA\\\";\\n}\\n.json-formatter-dark.json-formatter-row > a > .json-formatter-preview-text {\\n  opacity: 0;\\n  transition: opacity 0.15s ease-in;\\n  font-style: italic;\\n}\\n.json-formatter-dark.json-formatter-row:hover > a > .json-formatter-preview-text {\\n  opacity: 0.6;\\n}\\n.json-formatter-dark.json-formatter-row.json-formatter-open > .json-formatter-toggler-link .json-formatter-toggler:after {\\n  transform: rotate(90deg);\\n}\\n.json-formatter-dark.json-formatter-row.json-formatter-open > .json-formatter-children:after {\\n  display: inline-block;\\n}\\n.json-formatter-dark.json-formatter-row.json-formatter-open > a > .json-formatter-preview-text {\\n  display: none;\\n}\\n.json-formatter-dark.json-formatter-row.json-formatter-open.json-formatter-empty:after {\\n  display: block;\\n}\\n\");\n",
       "    },\n",
       "    function(Bokeh) {\n",
       "      inject_raw_css(\"table.panel-df {\\n    margin-left: auto;\\n    margin-right: auto;\\n    border: none;\\n    border-collapse: collapse;\\n    border-spacing: 0;\\n    color: black;\\n    font-size: 12px;\\n    table-layout: fixed;\\n    width: 100%;\\n}\\n\\n.panel-df tr, th, td {\\n    text-align: right;\\n    vertical-align: middle;\\n    padding: 0.5em 0.5em !important;\\n    line-height: normal;\\n    white-space: normal;\\n    max-width: none;\\n    border: none;\\n}\\n\\n.panel-df tbody {\\n    display: table-row-group;\\n    vertical-align: middle;\\n    border-color: inherit;\\n}\\n\\n.panel-df tbody tr:nth-child(odd) {\\n    background: #f5f5f5;\\n}\\n\\n.panel-df thead {\\n    border-bottom: 1px solid black;\\n    vertical-align: bottom;\\n}\\n\\n.panel-df tr:hover {\\n    background: lightblue !important;\\n    cursor: pointer;\\n}\\n\");\n",
       "    },\n",
       "    function(Bokeh) {\n",
       "      inject_raw_css(\".widget-box {\\n\\tmin-height: 20px;\\n\\tbackground-color: #f5f5f5;\\n\\tborder: 1px solid #e3e3e3 !important;\\n\\tborder-radius: 4px;\\n\\t-webkit-box-shadow: inset 0 1px 1px rgba(0,0,0,.05);\\n\\tbox-shadow: inset 0 1px 1px rgba(0,0,0,.05);\\n\\toverflow-x: hidden;\\n\\toverflow-y: hidden;\\n}\\n\\n.scrollable {\\n  overflow: scroll;\\n}\\n\\nprogress {\\n\\tappearance: none;\\n\\t-moz-appearance: none;\\n\\t-webkit-appearance: none;\\n\\n\\tborder: none;\\n\\theight: 20px;\\n\\tbackground-color: whiteSmoke;\\n\\tborder-radius: 3px;\\n\\tbox-shadow: 0 2px 3px rgba(0,0,0,.5) inset;\\n\\tcolor: royalblue;\\n\\tposition: relative;\\n\\tmargin: 0 0 1.5em;\\n}\\n\\nprogress[value]::-webkit-progress-bar {\\n\\tbackground-color: whiteSmoke;\\n\\tborder-radius: 3px;\\n\\tbox-shadow: 0 2px 3px rgba(0,0,0,.5) inset;\\n}\\n\\nprogress[value]::-webkit-progress-value {\\n\\tposition: relative;\\n\\n\\tbackground-size: 35px 20px, 100% 100%, 100% 100%;\\n\\tborder-radius:3px;\\n}\\n\\nprogress.active:not([value])::before {\\n\\tbackground-position: 10%;\\n\\tanimation-name: stripes;\\n\\tanimation-duration: 3s;\\n\\tanimation-timing-function: linear;\\n\\tanimation-iteration-count: infinite;\\n}\\n\\nprogress[value]::-moz-progress-bar {\\n\\tbackground-size: 35px 20px, 100% 100%, 100% 100%;\\n\\tborder-radius:3px;\\n}\\n\\nprogress:not([value])::-moz-progress-bar {\\n\\tborder-radius:3px;\\n\\tbackground:\\n\\tlinear-gradient(-45deg, transparent 33%, rgba(0, 0, 0, 0.2) 33%, rgba(0, 0, 0, 0.2) 66%, transparent 66%) left/2.5em 1.5em;\\n\\n}\\n\\nprogress.active:not([value])::-moz-progress-bar {\\n\\tbackground-position: 10%;\\n\\tanimation-name: stripes;\\n\\tanimation-duration: 3s;\\n\\tanimation-timing-function: linear;\\n\\tanimation-iteration-count: infinite;\\n}\\n\\nprogress.active:not([value])::-webkit-progress-bar {\\n\\tbackground-position: 10%;\\n\\tanimation-name: stripes;\\n\\tanimation-duration: 3s;\\n\\tanimation-timing-function: linear;\\n\\tanimation-iteration-count: infinite;\\n}\\n\\nprogress.primary[value]::-webkit-progress-value { background-color: #007bff; }\\nprogress.primary:not([value])::before { background-color: #007bff; }\\nprogress.primary:not([value])::-webkit-progress-bar { background-color: #007bff; }\\nprogress.primary::-moz-progress-bar { background-color: #007bff; }\\n\\nprogress.secondary[value]::-webkit-progress-value { background-color: #6c757d; }\\nprogress.secondary:not([value])::before { background-color: #6c757d; }\\nprogress.secondary:not([value])::-webkit-progress-bar { background-color: #6c757d; }\\nprogress.secondary::-moz-progress-bar { background-color: #6c757d; }\\n\\nprogress.success[value]::-webkit-progress-value { background-color: #28a745; }\\nprogress.success:not([value])::before { background-color: #28a745; }\\nprogress.success:not([value])::-webkit-progress-bar { background-color: #28a745; }\\nprogress.success::-moz-progress-bar { background-color: #28a745; }\\n\\nprogress.danger[value]::-webkit-progress-value { background-color: #dc3545; }\\nprogress.danger:not([value])::before { background-color: #dc3545; }\\nprogress.danger:not([value])::-webkit-progress-bar { background-color: #dc3545; }\\nprogress.danger::-moz-progress-bar { background-color: #dc3545; }\\n\\nprogress.warning[value]::-webkit-progress-value { background-color: #ffc107; }\\nprogress.warning:not([value])::before { background-color: #ffc107; }\\nprogress.warning:not([value])::-webkit-progress-bar { background-color: #ffc107; }\\nprogress.warning::-moz-progress-bar { background-color: #ffc107; }\\n\\nprogress.info[value]::-webkit-progress-value { background-color: #17a2b8; }\\nprogress.info:not([value])::before { background-color: #17a2b8; }\\nprogress.info:not([value])::-webkit-progress-bar { background-color: #17a2b8; }\\nprogress.info::-moz-progress-bar { background-color: #17a2b8; }\\n\\nprogress.light[value]::-webkit-progress-value { background-color: #f8f9fa; }\\nprogress.light:not([value])::before { background-color: #f8f9fa; }\\nprogress.light:not([value])::-webkit-progress-bar { background-color: #f8f9fa; }\\nprogress.light::-moz-progress-bar { background-color: #f8f9fa; }\\n\\nprogress.dark[value]::-webkit-progress-value { background-color: #343a40; }\\nprogress.dark:not([value])::-webkit-progress-bar { background-color: #343a40; }\\nprogress.dark:not([value])::before { background-color: #343a40; }\\nprogress.dark::-moz-progress-bar { background-color: #343a40; }\\n\\nprogress:not([value])::-webkit-progress-bar {\\n\\tborder-radius: 3px;\\n\\tbackground:\\n\\tlinear-gradient(-45deg, transparent 33%, rgba(0, 0, 0, 0.2) 33%, rgba(0, 0, 0, 0.2) 66%, transparent 66%) left/2.5em 1.5em;\\n}\\nprogress:not([value])::before {\\n\\tcontent:\\\" \\\";\\n\\tposition:absolute;\\n\\theight: 20px;\\n\\ttop:0;\\n\\tleft:0;\\n\\tright:0;\\n\\tbottom:0;\\n\\tborder-radius: 3px;\\n\\tbackground:\\n\\tlinear-gradient(-45deg, transparent 33%, rgba(0, 0, 0, 0.2) 33%, rgba(0, 0, 0, 0.2) 66%, transparent 66%) left/2.5em 1.5em;\\n}\\n\\n@keyframes stripes {\\n  from {background-position: 0%}\\n  to {background-position: 100%}\\n}\");\n",
       "    },\n",
       "    function(Bokeh) {\n",
       "      inject_raw_css(\".codehilite .hll { background-color: #ffffcc }\\n.codehilite  { background: #f8f8f8; }\\n.codehilite .c { color: #408080; font-style: italic } /* Comment */\\n.codehilite .err { border: 1px solid #FF0000 } /* Error */\\n.codehilite .k { color: #008000; font-weight: bold } /* Keyword */\\n.codehilite .o { color: #666666 } /* Operator */\\n.codehilite .ch { color: #408080; font-style: italic } /* Comment.Hashbang */\\n.codehilite .cm { color: #408080; font-style: italic } /* Comment.Multiline */\\n.codehilite .cp { color: #BC7A00 } /* Comment.Preproc */\\n.codehilite .cpf { color: #408080; font-style: italic } /* Comment.PreprocFile */\\n.codehilite .c1 { color: #408080; font-style: italic } /* Comment.Single */\\n.codehilite .cs { color: #408080; font-style: italic } /* Comment.Special */\\n.codehilite .gd { color: #A00000 } /* Generic.Deleted */\\n.codehilite .ge { font-style: italic } /* Generic.Emph */\\n.codehilite .gr { color: #FF0000 } /* Generic.Error */\\n.codehilite .gh { color: #000080; font-weight: bold } /* Generic.Heading */\\n.codehilite .gi { color: #00A000 } /* Generic.Inserted */\\n.codehilite .go { color: #888888 } /* Generic.Output */\\n.codehilite .gp { color: #000080; font-weight: bold } /* Generic.Prompt */\\n.codehilite .gs { font-weight: bold } /* Generic.Strong */\\n.codehilite .gu { color: #800080; font-weight: bold } /* Generic.Subheading */\\n.codehilite .gt { color: #0044DD } /* Generic.Traceback */\\n.codehilite .kc { color: #008000; font-weight: bold } /* Keyword.Constant */\\n.codehilite .kd { color: #008000; font-weight: bold } /* Keyword.Declaration */\\n.codehilite .kn { color: #008000; font-weight: bold } /* Keyword.Namespace */\\n.codehilite .kp { color: #008000 } /* Keyword.Pseudo */\\n.codehilite .kr { color: #008000; font-weight: bold } /* Keyword.Reserved */\\n.codehilite .kt { color: #B00040 } /* Keyword.Type */\\n.codehilite .m { color: #666666 } /* Literal.Number */\\n.codehilite .s { color: #BA2121 } /* Literal.String */\\n.codehilite .na { color: #7D9029 } /* Name.Attribute */\\n.codehilite .nb { color: #008000 } /* Name.Builtin */\\n.codehilite .nc { color: #0000FF; font-weight: bold } /* Name.Class */\\n.codehilite .no { color: #880000 } /* Name.Constant */\\n.codehilite .nd { color: #AA22FF } /* Name.Decorator */\\n.codehilite .ni { color: #999999; font-weight: bold } /* Name.Entity */\\n.codehilite .ne { color: #D2413A; font-weight: bold } /* Name.Exception */\\n.codehilite .nf { color: #0000FF } /* Name.Function */\\n.codehilite .nl { color: #A0A000 } /* Name.Label */\\n.codehilite .nn { color: #0000FF; font-weight: bold } /* Name.Namespace */\\n.codehilite .nt { color: #008000; font-weight: bold } /* Name.Tag */\\n.codehilite .nv { color: #19177C } /* Name.Variable */\\n.codehilite .ow { color: #AA22FF; font-weight: bold } /* Operator.Word */\\n.codehilite .w { color: #bbbbbb } /* Text.Whitespace */\\n.codehilite .mb { color: #666666 } /* Literal.Number.Bin */\\n.codehilite .mf { color: #666666 } /* Literal.Number.Float */\\n.codehilite .mh { color: #666666 } /* Literal.Number.Hex */\\n.codehilite .mi { color: #666666 } /* Literal.Number.Integer */\\n.codehilite .mo { color: #666666 } /* Literal.Number.Oct */\\n.codehilite .sa { color: #BA2121 } /* Literal.String.Affix */\\n.codehilite .sb { color: #BA2121 } /* Literal.String.Backtick */\\n.codehilite .sc { color: #BA2121 } /* Literal.String.Char */\\n.codehilite .dl { color: #BA2121 } /* Literal.String.Delimiter */\\n.codehilite .sd { color: #BA2121; font-style: italic } /* Literal.String.Doc */\\n.codehilite .s2 { color: #BA2121 } /* Literal.String.Double */\\n.codehilite .se { color: #BB6622; font-weight: bold } /* Literal.String.Escape */\\n.codehilite .sh { color: #BA2121 } /* Literal.String.Heredoc */\\n.codehilite .si { color: #BB6688; font-weight: bold } /* Literal.String.Interpol */\\n.codehilite .sx { color: #008000 } /* Literal.String.Other */\\n.codehilite .sr { color: #BB6688 } /* Literal.String.Regex */\\n.codehilite .s1 { color: #BA2121 } /* Literal.String.Single */\\n.codehilite .ss { color: #19177C } /* Literal.String.Symbol */\\n.codehilite .bp { color: #008000 } /* Name.Builtin.Pseudo */\\n.codehilite .fm { color: #0000FF } /* Name.Function.Magic */\\n.codehilite .vc { color: #19177C } /* Name.Variable.Class */\\n.codehilite .vg { color: #19177C } /* Name.Variable.Global */\\n.codehilite .vi { color: #19177C } /* Name.Variable.Instance */\\n.codehilite .vm { color: #19177C } /* Name.Variable.Magic */\\n.codehilite .il { color: #666666 } /* Literal.Number.Integer.Long */\\n\\n.markdown h1 { margin-block-start: 0.34em }\\n.markdown h2 { margin-block-start: 0.42em }\\n.markdown h3 { margin-block-start: 0.5em }\\n.markdown h4 { margin-block-start: 0.67em }\\n.markdown h5 { margin-block-start: 0.84em }\\n.markdown h6 { margin-block-start: 1.17em }\\n.markdown ul { padding-inline-start: 2em }\\n.markdown ol { padding-inline-start: 2em }\\n.markdown strong { font-weight: 600 }\\n.markdown a { color: -webkit-link }\\n.markdown a { color: -moz-hyperlinkText }\\n\");\n",
       "    },\n",
       "    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "    function(Bokeh) {\n",
       "    \n",
       "    \n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    \n",
       "    if (root.Bokeh !== undefined || force === true) {\n",
       "      \n",
       "    for (var i = 0; i < inline_js.length; i++) {\n",
       "      inline_js[i].call(root, root.Bokeh);\n",
       "    }\n",
       "    if (force === true) {\n",
       "        display_loaded();\n",
       "      }} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      var cell = $(document.getElementById(\"1467\")).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  var force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n  \n\n  \n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  var NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    var el = document.getElementById(\"1467\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error() {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (var i = 0; i < css_urls.length; i++) {\n      var url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    const hashes = {\"https://cdn.bokeh.org/bokeh/release/bokeh-2.0.1.min.js\": \"JpP8FXbgAZLkfur7LiK3j9AGBhHNIvF742meBJrjO2ShJDhCG2I1uVvW+0DUtrmc\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.0.1.min.js\": \"xZlADit0Q04ISQEdKg2k3L4W9AwQBAuDs9nJL9fM/WwzL1tEU9VPNezOFX0nLEAz\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.0.1.min.js\": \"4BuPRZkdMKSnj3zoxiNrQ86XgNw0rYmBOxe7nshquXwwcauupgBF2DHLVG1WuZlV\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.0.1.min.js\": \"Dv1SQ87hmDqK6S5OhBf0bCuwAEvL5QYL0PuR/F1SPVhCS/r/abjkbpKDYL2zeM19\"};\n\n    for (var i = 0; i < js_urls.length; i++) {\n      var url = js_urls[i];\n      var element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.async = false;\n      element.src = url;\n      if (url in hashes) {\n        element.crossOrigin = \"anonymous\";\n        element.integrity = \"sha384-\" + hashes[url];\n      }\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };var element = document.getElementById(\"1467\");\n  if (element == null) {\n    console.error(\"Bokeh: ERROR: autoload.js configured with elementid '1467' but no matching script tag was found. \")\n    return false;\n  }\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  \n  var js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.0.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.0.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.0.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.0.1.min.js\", \"https://unpkg.com/@holoviz/panel@^0.9.5/dist/panel.min.js\"];\n  var css_urls = [];\n  \n\n  var inline_js = [\n    function(Bokeh) {\n      inject_raw_css(\".json-formatter-row {\\n  font-family: monospace;\\n}\\n.json-formatter-row,\\n.json-formatter-row a,\\n.json-formatter-row a:hover {\\n  color: black;\\n  text-decoration: none;\\n}\\n.json-formatter-row .json-formatter-row {\\n  margin-left: 1rem;\\n}\\n.json-formatter-row .json-formatter-children.json-formatter-empty {\\n  opacity: 0.5;\\n  margin-left: 1rem;\\n}\\n.json-formatter-row .json-formatter-children.json-formatter-empty:after {\\n  display: none;\\n}\\n.json-formatter-row .json-formatter-children.json-formatter-empty.json-formatter-object:after {\\n  content: \\\"No properties\\\";\\n}\\n.json-formatter-row .json-formatter-children.json-formatter-empty.json-formatter-array:after {\\n  content: \\\"[]\\\";\\n}\\n.json-formatter-row .json-formatter-string,\\n.json-formatter-row .json-formatter-stringifiable {\\n  color: green;\\n  white-space: pre;\\n  word-wrap: break-word;\\n}\\n.json-formatter-row .json-formatter-number {\\n  color: blue;\\n}\\n.json-formatter-row .json-formatter-boolean {\\n  color: red;\\n}\\n.json-formatter-row .json-formatter-null {\\n  color: #855A00;\\n}\\n.json-formatter-row .json-formatter-undefined {\\n  color: #ca0b69;\\n}\\n.json-formatter-row .json-formatter-function {\\n  color: #FF20ED;\\n}\\n.json-formatter-row .json-formatter-date {\\n  background-color: rgba(0, 0, 0, 0.05);\\n}\\n.json-formatter-row .json-formatter-url {\\n  text-decoration: underline;\\n  color: blue;\\n  cursor: pointer;\\n}\\n.json-formatter-row .json-formatter-bracket {\\n  color: blue;\\n}\\n.json-formatter-row .json-formatter-key {\\n  color: #00008B;\\n  padding-right: 0.2rem;\\n}\\n.json-formatter-row .json-formatter-toggler-link {\\n  cursor: pointer;\\n}\\n.json-formatter-row .json-formatter-toggler {\\n  line-height: 1.2rem;\\n  font-size: 0.7rem;\\n  vertical-align: middle;\\n  opacity: 0.6;\\n  cursor: pointer;\\n  padding-right: 0.2rem;\\n}\\n.json-formatter-row .json-formatter-toggler:after {\\n  display: inline-block;\\n  transition: transform 100ms ease-in;\\n  content: \\\"\\\\25BA\\\";\\n}\\n.json-formatter-row > a > .json-formatter-preview-text {\\n  opacity: 0;\\n  transition: opacity 0.15s ease-in;\\n  font-style: italic;\\n}\\n.json-formatter-row:hover > a > .json-formatter-preview-text {\\n  opacity: 0.6;\\n}\\n.json-formatter-row.json-formatter-open > .json-formatter-toggler-link .json-formatter-toggler:after {\\n  transform: rotate(90deg);\\n}\\n.json-formatter-row.json-formatter-open > .json-formatter-children:after {\\n  display: inline-block;\\n}\\n.json-formatter-row.json-formatter-open > a > .json-formatter-preview-text {\\n  display: none;\\n}\\n.json-formatter-row.json-formatter-open.json-formatter-empty:after {\\n  display: block;\\n}\\n.json-formatter-dark.json-formatter-row {\\n  font-family: monospace;\\n}\\n.json-formatter-dark.json-formatter-row,\\n.json-formatter-dark.json-formatter-row a,\\n.json-formatter-dark.json-formatter-row a:hover {\\n  color: white;\\n  text-decoration: none;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-row {\\n  margin-left: 1rem;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-children.json-formatter-empty {\\n  opacity: 0.5;\\n  margin-left: 1rem;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-children.json-formatter-empty:after {\\n  display: none;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-children.json-formatter-empty.json-formatter-object:after {\\n  content: \\\"No properties\\\";\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-children.json-formatter-empty.json-formatter-array:after {\\n  content: \\\"[]\\\";\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-string,\\n.json-formatter-dark.json-formatter-row .json-formatter-stringifiable {\\n  color: #31F031;\\n  white-space: pre;\\n  word-wrap: break-word;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-number {\\n  color: #66C2FF;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-boolean {\\n  color: #EC4242;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-null {\\n  color: #EEC97D;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-undefined {\\n  color: #ef8fbe;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-function {\\n  color: #FD48CB;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-date {\\n  background-color: rgba(255, 255, 255, 0.05);\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-url {\\n  text-decoration: underline;\\n  color: #027BFF;\\n  cursor: pointer;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-bracket {\\n  color: #9494FF;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-key {\\n  color: #23A0DB;\\n  padding-right: 0.2rem;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-toggler-link {\\n  cursor: pointer;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-toggler {\\n  line-height: 1.2rem;\\n  font-size: 0.7rem;\\n  vertical-align: middle;\\n  opacity: 0.6;\\n  cursor: pointer;\\n  padding-right: 0.2rem;\\n}\\n.json-formatter-dark.json-formatter-row .json-formatter-toggler:after {\\n  display: inline-block;\\n  transition: transform 100ms ease-in;\\n  content: \\\"\\\\25BA\\\";\\n}\\n.json-formatter-dark.json-formatter-row > a > .json-formatter-preview-text {\\n  opacity: 0;\\n  transition: opacity 0.15s ease-in;\\n  font-style: italic;\\n}\\n.json-formatter-dark.json-formatter-row:hover > a > .json-formatter-preview-text {\\n  opacity: 0.6;\\n}\\n.json-formatter-dark.json-formatter-row.json-formatter-open > .json-formatter-toggler-link .json-formatter-toggler:after {\\n  transform: rotate(90deg);\\n}\\n.json-formatter-dark.json-formatter-row.json-formatter-open > .json-formatter-children:after {\\n  display: inline-block;\\n}\\n.json-formatter-dark.json-formatter-row.json-formatter-open > a > .json-formatter-preview-text {\\n  display: none;\\n}\\n.json-formatter-dark.json-formatter-row.json-formatter-open.json-formatter-empty:after {\\n  display: block;\\n}\\n\");\n    },\n    function(Bokeh) {\n      inject_raw_css(\"table.panel-df {\\n    margin-left: auto;\\n    margin-right: auto;\\n    border: none;\\n    border-collapse: collapse;\\n    border-spacing: 0;\\n    color: black;\\n    font-size: 12px;\\n    table-layout: fixed;\\n    width: 100%;\\n}\\n\\n.panel-df tr, th, td {\\n    text-align: right;\\n    vertical-align: middle;\\n    padding: 0.5em 0.5em !important;\\n    line-height: normal;\\n    white-space: normal;\\n    max-width: none;\\n    border: none;\\n}\\n\\n.panel-df tbody {\\n    display: table-row-group;\\n    vertical-align: middle;\\n    border-color: inherit;\\n}\\n\\n.panel-df tbody tr:nth-child(odd) {\\n    background: #f5f5f5;\\n}\\n\\n.panel-df thead {\\n    border-bottom: 1px solid black;\\n    vertical-align: bottom;\\n}\\n\\n.panel-df tr:hover {\\n    background: lightblue !important;\\n    cursor: pointer;\\n}\\n\");\n    },\n    function(Bokeh) {\n      inject_raw_css(\".widget-box {\\n\\tmin-height: 20px;\\n\\tbackground-color: #f5f5f5;\\n\\tborder: 1px solid #e3e3e3 !important;\\n\\tborder-radius: 4px;\\n\\t-webkit-box-shadow: inset 0 1px 1px rgba(0,0,0,.05);\\n\\tbox-shadow: inset 0 1px 1px rgba(0,0,0,.05);\\n\\toverflow-x: hidden;\\n\\toverflow-y: hidden;\\n}\\n\\n.scrollable {\\n  overflow: scroll;\\n}\\n\\nprogress {\\n\\tappearance: none;\\n\\t-moz-appearance: none;\\n\\t-webkit-appearance: none;\\n\\n\\tborder: none;\\n\\theight: 20px;\\n\\tbackground-color: whiteSmoke;\\n\\tborder-radius: 3px;\\n\\tbox-shadow: 0 2px 3px rgba(0,0,0,.5) inset;\\n\\tcolor: royalblue;\\n\\tposition: relative;\\n\\tmargin: 0 0 1.5em;\\n}\\n\\nprogress[value]::-webkit-progress-bar {\\n\\tbackground-color: whiteSmoke;\\n\\tborder-radius: 3px;\\n\\tbox-shadow: 0 2px 3px rgba(0,0,0,.5) inset;\\n}\\n\\nprogress[value]::-webkit-progress-value {\\n\\tposition: relative;\\n\\n\\tbackground-size: 35px 20px, 100% 100%, 100% 100%;\\n\\tborder-radius:3px;\\n}\\n\\nprogress.active:not([value])::before {\\n\\tbackground-position: 10%;\\n\\tanimation-name: stripes;\\n\\tanimation-duration: 3s;\\n\\tanimation-timing-function: linear;\\n\\tanimation-iteration-count: infinite;\\n}\\n\\nprogress[value]::-moz-progress-bar {\\n\\tbackground-size: 35px 20px, 100% 100%, 100% 100%;\\n\\tborder-radius:3px;\\n}\\n\\nprogress:not([value])::-moz-progress-bar {\\n\\tborder-radius:3px;\\n\\tbackground:\\n\\tlinear-gradient(-45deg, transparent 33%, rgba(0, 0, 0, 0.2) 33%, rgba(0, 0, 0, 0.2) 66%, transparent 66%) left/2.5em 1.5em;\\n\\n}\\n\\nprogress.active:not([value])::-moz-progress-bar {\\n\\tbackground-position: 10%;\\n\\tanimation-name: stripes;\\n\\tanimation-duration: 3s;\\n\\tanimation-timing-function: linear;\\n\\tanimation-iteration-count: infinite;\\n}\\n\\nprogress.active:not([value])::-webkit-progress-bar {\\n\\tbackground-position: 10%;\\n\\tanimation-name: stripes;\\n\\tanimation-duration: 3s;\\n\\tanimation-timing-function: linear;\\n\\tanimation-iteration-count: infinite;\\n}\\n\\nprogress.primary[value]::-webkit-progress-value { background-color: #007bff; }\\nprogress.primary:not([value])::before { background-color: #007bff; }\\nprogress.primary:not([value])::-webkit-progress-bar { background-color: #007bff; }\\nprogress.primary::-moz-progress-bar { background-color: #007bff; }\\n\\nprogress.secondary[value]::-webkit-progress-value { background-color: #6c757d; }\\nprogress.secondary:not([value])::before { background-color: #6c757d; }\\nprogress.secondary:not([value])::-webkit-progress-bar { background-color: #6c757d; }\\nprogress.secondary::-moz-progress-bar { background-color: #6c757d; }\\n\\nprogress.success[value]::-webkit-progress-value { background-color: #28a745; }\\nprogress.success:not([value])::before { background-color: #28a745; }\\nprogress.success:not([value])::-webkit-progress-bar { background-color: #28a745; }\\nprogress.success::-moz-progress-bar { background-color: #28a745; }\\n\\nprogress.danger[value]::-webkit-progress-value { background-color: #dc3545; }\\nprogress.danger:not([value])::before { background-color: #dc3545; }\\nprogress.danger:not([value])::-webkit-progress-bar { background-color: #dc3545; }\\nprogress.danger::-moz-progress-bar { background-color: #dc3545; }\\n\\nprogress.warning[value]::-webkit-progress-value { background-color: #ffc107; }\\nprogress.warning:not([value])::before { background-color: #ffc107; }\\nprogress.warning:not([value])::-webkit-progress-bar { background-color: #ffc107; }\\nprogress.warning::-moz-progress-bar { background-color: #ffc107; }\\n\\nprogress.info[value]::-webkit-progress-value { background-color: #17a2b8; }\\nprogress.info:not([value])::before { background-color: #17a2b8; }\\nprogress.info:not([value])::-webkit-progress-bar { background-color: #17a2b8; }\\nprogress.info::-moz-progress-bar { background-color: #17a2b8; }\\n\\nprogress.light[value]::-webkit-progress-value { background-color: #f8f9fa; }\\nprogress.light:not([value])::before { background-color: #f8f9fa; }\\nprogress.light:not([value])::-webkit-progress-bar { background-color: #f8f9fa; }\\nprogress.light::-moz-progress-bar { background-color: #f8f9fa; }\\n\\nprogress.dark[value]::-webkit-progress-value { background-color: #343a40; }\\nprogress.dark:not([value])::-webkit-progress-bar { background-color: #343a40; }\\nprogress.dark:not([value])::before { background-color: #343a40; }\\nprogress.dark::-moz-progress-bar { background-color: #343a40; }\\n\\nprogress:not([value])::-webkit-progress-bar {\\n\\tborder-radius: 3px;\\n\\tbackground:\\n\\tlinear-gradient(-45deg, transparent 33%, rgba(0, 0, 0, 0.2) 33%, rgba(0, 0, 0, 0.2) 66%, transparent 66%) left/2.5em 1.5em;\\n}\\nprogress:not([value])::before {\\n\\tcontent:\\\" \\\";\\n\\tposition:absolute;\\n\\theight: 20px;\\n\\ttop:0;\\n\\tleft:0;\\n\\tright:0;\\n\\tbottom:0;\\n\\tborder-radius: 3px;\\n\\tbackground:\\n\\tlinear-gradient(-45deg, transparent 33%, rgba(0, 0, 0, 0.2) 33%, rgba(0, 0, 0, 0.2) 66%, transparent 66%) left/2.5em 1.5em;\\n}\\n\\n@keyframes stripes {\\n  from {background-position: 0%}\\n  to {background-position: 100%}\\n}\");\n    },\n    function(Bokeh) {\n      inject_raw_css(\".codehilite .hll { background-color: #ffffcc }\\n.codehilite  { background: #f8f8f8; }\\n.codehilite .c { color: #408080; font-style: italic } /* Comment */\\n.codehilite .err { border: 1px solid #FF0000 } /* Error */\\n.codehilite .k { color: #008000; font-weight: bold } /* Keyword */\\n.codehilite .o { color: #666666 } /* Operator */\\n.codehilite .ch { color: #408080; font-style: italic } /* Comment.Hashbang */\\n.codehilite .cm { color: #408080; font-style: italic } /* Comment.Multiline */\\n.codehilite .cp { color: #BC7A00 } /* Comment.Preproc */\\n.codehilite .cpf { color: #408080; font-style: italic } /* Comment.PreprocFile */\\n.codehilite .c1 { color: #408080; font-style: italic } /* Comment.Single */\\n.codehilite .cs { color: #408080; font-style: italic } /* Comment.Special */\\n.codehilite .gd { color: #A00000 } /* Generic.Deleted */\\n.codehilite .ge { font-style: italic } /* Generic.Emph */\\n.codehilite .gr { color: #FF0000 } /* Generic.Error */\\n.codehilite .gh { color: #000080; font-weight: bold } /* Generic.Heading */\\n.codehilite .gi { color: #00A000 } /* Generic.Inserted */\\n.codehilite .go { color: #888888 } /* Generic.Output */\\n.codehilite .gp { color: #000080; font-weight: bold } /* Generic.Prompt */\\n.codehilite .gs { font-weight: bold } /* Generic.Strong */\\n.codehilite .gu { color: #800080; font-weight: bold } /* Generic.Subheading */\\n.codehilite .gt { color: #0044DD } /* Generic.Traceback */\\n.codehilite .kc { color: #008000; font-weight: bold } /* Keyword.Constant */\\n.codehilite .kd { color: #008000; font-weight: bold } /* Keyword.Declaration */\\n.codehilite .kn { color: #008000; font-weight: bold } /* Keyword.Namespace */\\n.codehilite .kp { color: #008000 } /* Keyword.Pseudo */\\n.codehilite .kr { color: #008000; font-weight: bold } /* Keyword.Reserved */\\n.codehilite .kt { color: #B00040 } /* Keyword.Type */\\n.codehilite .m { color: #666666 } /* Literal.Number */\\n.codehilite .s { color: #BA2121 } /* Literal.String */\\n.codehilite .na { color: #7D9029 } /* Name.Attribute */\\n.codehilite .nb { color: #008000 } /* Name.Builtin */\\n.codehilite .nc { color: #0000FF; font-weight: bold } /* Name.Class */\\n.codehilite .no { color: #880000 } /* Name.Constant */\\n.codehilite .nd { color: #AA22FF } /* Name.Decorator */\\n.codehilite .ni { color: #999999; font-weight: bold } /* Name.Entity */\\n.codehilite .ne { color: #D2413A; font-weight: bold } /* Name.Exception */\\n.codehilite .nf { color: #0000FF } /* Name.Function */\\n.codehilite .nl { color: #A0A000 } /* Name.Label */\\n.codehilite .nn { color: #0000FF; font-weight: bold } /* Name.Namespace */\\n.codehilite .nt { color: #008000; font-weight: bold } /* Name.Tag */\\n.codehilite .nv { color: #19177C } /* Name.Variable */\\n.codehilite .ow { color: #AA22FF; font-weight: bold } /* Operator.Word */\\n.codehilite .w { color: #bbbbbb } /* Text.Whitespace */\\n.codehilite .mb { color: #666666 } /* Literal.Number.Bin */\\n.codehilite .mf { color: #666666 } /* Literal.Number.Float */\\n.codehilite .mh { color: #666666 } /* Literal.Number.Hex */\\n.codehilite .mi { color: #666666 } /* Literal.Number.Integer */\\n.codehilite .mo { color: #666666 } /* Literal.Number.Oct */\\n.codehilite .sa { color: #BA2121 } /* Literal.String.Affix */\\n.codehilite .sb { color: #BA2121 } /* Literal.String.Backtick */\\n.codehilite .sc { color: #BA2121 } /* Literal.String.Char */\\n.codehilite .dl { color: #BA2121 } /* Literal.String.Delimiter */\\n.codehilite .sd { color: #BA2121; font-style: italic } /* Literal.String.Doc */\\n.codehilite .s2 { color: #BA2121 } /* Literal.String.Double */\\n.codehilite .se { color: #BB6622; font-weight: bold } /* Literal.String.Escape */\\n.codehilite .sh { color: #BA2121 } /* Literal.String.Heredoc */\\n.codehilite .si { color: #BB6688; font-weight: bold } /* Literal.String.Interpol */\\n.codehilite .sx { color: #008000 } /* Literal.String.Other */\\n.codehilite .sr { color: #BB6688 } /* Literal.String.Regex */\\n.codehilite .s1 { color: #BA2121 } /* Literal.String.Single */\\n.codehilite .ss { color: #19177C } /* Literal.String.Symbol */\\n.codehilite .bp { color: #008000 } /* Name.Builtin.Pseudo */\\n.codehilite .fm { color: #0000FF } /* Name.Function.Magic */\\n.codehilite .vc { color: #19177C } /* Name.Variable.Class */\\n.codehilite .vg { color: #19177C } /* Name.Variable.Global */\\n.codehilite .vi { color: #19177C } /* Name.Variable.Instance */\\n.codehilite .vm { color: #19177C } /* Name.Variable.Magic */\\n.codehilite .il { color: #666666 } /* Literal.Number.Integer.Long */\\n\\n.markdown h1 { margin-block-start: 0.34em }\\n.markdown h2 { margin-block-start: 0.42em }\\n.markdown h3 { margin-block-start: 0.5em }\\n.markdown h4 { margin-block-start: 0.67em }\\n.markdown h5 { margin-block-start: 0.84em }\\n.markdown h6 { margin-block-start: 1.17em }\\n.markdown ul { padding-inline-start: 2em }\\n.markdown ol { padding-inline-start: 2em }\\n.markdown strong { font-weight: 600 }\\n.markdown a { color: -webkit-link }\\n.markdown a { color: -moz-hyperlinkText }\\n\");\n    },\n    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\n    function(Bokeh) {\n    \n    \n    }\n  ];\n\n  function run_inline_js() {\n    \n    if (root.Bokeh !== undefined || force === true) {\n      \n    for (var i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n    if (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      var cell = $(document.getElementById(\"1467\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# See labels of UMAP projection\n",
    "import umap.plot\n",
    "#umap.plot.points(mapper)\n",
    "umap.plot.output_notebook()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alongside the data point, we create hover data, consisting of the genome’s association, cluster number, taxonomy, and bray-curtis similarity (with reference being the first genome in the entire dataset)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.series.Series'>\n",
      "      distance                                              taxon     site\n",
      "0          1.0  [Bacteroidetes, Bacteroidetes, Cytophagia, Cyt...  Unknown\n",
      "1          0.0  [Alphaproteobacteria, Proteobacteria, Alphapro...  Unknown\n",
      "2          0.0  [Alphaproteobacteria, Proteobacteria, Alphapro...  Unknown\n",
      "3          0.0  [Burkholderiales, Proteobacteria, Betaproteoba...  Unknown\n",
      "4          0.0  [Pseudomonas, Proteobacteria, Gammaproteobacte...  Unknown\n",
      "...        ...                                                ...      ...\n",
      "3832       0.0  [Alphaproteobacteria, Proteobacteria, Alphapro...  Cologne\n",
      "3833       0.0  [Alphaproteobacteria, Proteobacteria, Alphapro...  Cologne\n",
      "3834       0.0  [Alphaproteobacteria, Proteobacteria, Alphapro...  Cologne\n",
      "3835       0.0  [Actinobacteria_one, Actinobacteria, Actinobac...  Cologne\n",
      "3836       0.0  [Actinobacteria_one, Actinobacteria, Actinobac...  Cologne\n",
      "\n",
      "[3837 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "print(type(site))\n",
    "hover_data = pd.DataFrame({'distance':distance.to_numpy(), 'taxon': taxon.to_numpy(), 'site': site.to_numpy()})\n",
    "print(hover_data)\n",
    "#p = umap.plot.interactive(mapper,hover_data=hover_data, point_size=2)\n",
    "#umap.plot.show(p)\n",
    "hover_data['association'] = list(association)\n",
    "hover_data['cluster'] = clusterer.labels_\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot the clusters with the hover data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "print(hover_data.head())\n",
    "p = umap.plot.interactive(mapper, labels = clusterer.labels_, color_key_cmap='Paired', background='black',hover_data=hover_data, point_size=15)\n",
    "umap.plot.show(p)\n",
    "# Colour palette for HDBScan Plot\n",
    "#palette = sns.color_palette()\n",
    "#color_palette = sns.color_palette('Paired', clusterer.labels_.max()+1)\n",
    "#cluster_colors = [color_palette[x] if x >= 0 else (0.5, 0.5, 0.5) for x in clusterer.labels_]\n",
    "#cluster_member_colors = [sns.desaturate(x, p) for x, p in\n",
    "                        # zip(cluster_colors, clusterer.probabilities_)]\n",
    "#plt.scatter(*embedding.T, s=5, linewidth=0, c=cluster_member_colors, alpha=0.25)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A manual check of the hover data for each cluster shows great taxonomic similarity both by bray-curtis similarity, as well as closeness in taxonomy (similarity down to the genus). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cluster:  0\n",
      "Cluster:  1\n",
      "Cluster:  2\n",
      "Cluster:  3\n",
      "Cluster:  4\n",
      "Cluster:  5\n",
      "Cluster:  6\n",
      "Cluster:  7\n",
      "Cluster:  8\n",
      "Cluster:  9\n",
      "Cluster:  10\n",
      "Cluster:  11\n",
      "Cluster:  12\n",
      "Cluster:  13\n",
      "Cluster:  14\n",
      "Cluster:  15\n",
      "Cluster:  16\n",
      "Cluster:  17\n",
      "Cluster:  18\n",
      "Cluster:  19\n",
      "Cluster:  20\n",
      "Cluster:  21\n",
      "Cluster:  22\n",
      "Cluster:  23\n",
      "Cluster:  24\n",
      "Cluster:  25\n",
      "Cluster:  26\n",
      "Cluster:  27\n",
      "Cluster:  28\n",
      "Cluster:  29\n",
      "Cluster:  30\n",
      "Cluster:  31\n",
      "Cluster:  32\n",
      "Cluster:  33\n",
      "Cluster:  34\n",
      "Cluster:  35\n",
      "Cluster:  36\n",
      "Cluster:  37\n",
      "Cluster:  38\n",
      "Cluster:  39\n",
      "Cluster:  40\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "    Test to see which clusters match up to what taxonomy. Print statement is commented out to avoid clutter.\n",
    "'''\n",
    "for clusternum in range(0, clusterer.labels_.max()+1):\n",
    "    print('Cluster: ', clusternum)\n",
    "    for genome in range(len(hover_data)):\n",
    "        if hover_data.iloc[genome, -1] == clusternum:\n",
    "            #print(hover_data.iloc[genome, taxoncol])\n",
    "            pass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It seems a good proportion of the data is clustered as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Proportion of data clustered:  0.9051342194422726\n"
     ]
    }
   ],
   "source": [
    "labels = hdbscan.HDBSCAN(\n",
    "    min_samples=40,\n",
    "    min_cluster_size=10,\n",
    ").fit_predict(embedding)\n",
    "print('Proportion of data clustered: ', np.sum(labels >= 0) / vdf.shape[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Having validated our clusters, we move on to isolate genes that are unique to each cluster. Using this approach, we can find families of genes that are specific to groups of bacteria and the symbiotic relationships they may have."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We first create an empty matrix indexed by association: “PA”, “NPA”, “soil”, for plant, non-plant, and soil-association bacteria respectively. Column names range from 0-39, being the cluster labels we have created. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "metadata": {},
   "outputs": [],
   "source": [
    "emptyrow = [[] for i in range(clusterer.labels_.max() + 2)]\n",
    "gene_clusters = pd.DataFrame(data = [emptyrow for i in range(3)], index = ['PA', 'NPA', 'soil'], \n",
    "                             columns = list(range(clusterer.labels_.min(), clusterer.labels_.max()+1)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each cluster should have some NPA genes and PA genes:\n",
    "\n",
    "Find unique clusters, find all genes of NPAS and all genes of PAs, find unique genes to each.\n",
    "\n",
    "For each classification, get a list of the genes. Add to cluster under NPA or PA. \n",
    "Then, for each cluster, iterate through each NPA gene and check if it is a PA/soil gene. \n",
    "If it's unique to NPA, then add it to uniques."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then, we loop through our cluster labels and select one by one possible genes of interest for each genome, by returning a list of all genes present in the genome. By parsing our matrix containing the metadata for the association of that genome, we obtain an index. Similarly, we set our column name to the cluster label. By grouping together all possible genes of interest by the cluster label, we can find genes unique to each cluster. But first, we turn all the genes of interest into a set, as we are not interested in duplicate genes. We thus annotate our empty matrix with all the genes of interest in a certain cluster, with a certain association."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "'>' not supported between instances of 'str' and 'int'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-250-060a6a81ec40>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0;31m# check for genes of interest\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0mrow\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mvdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcluster\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m     \u001b[0mgenes_of_interest\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mrow\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m     \u001b[0mclassification\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0massociation\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcluster\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m     \u001b[0mtotal_genes\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgene_clusters\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mat\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mclassification\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mclusterer\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlabels_\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcluster\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mgenes_of_interest\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/nathan/miniconda3/envs/jhub/lib/python3.6/site-packages/pandas/core/ops/common.py\u001b[0m in \u001b[0;36mnew_method\u001b[0;34m(self, other)\u001b[0m\n\u001b[1;32m     62\u001b[0m         \u001b[0mother\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mitem_from_zerodim\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mother\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     63\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 64\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mother\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     65\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     66\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mnew_method\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/nathan/miniconda3/envs/jhub/lib/python3.6/site-packages/pandas/core/ops/__init__.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(self, other)\u001b[0m\n\u001b[1;32m    527\u001b[0m         \u001b[0mrvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mextract_array\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mother\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mextract_numpy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    528\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 529\u001b[0;31m         \u001b[0mres_values\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcomparison_op\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mop\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    530\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    531\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0m_construct_result\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mres_values\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mres_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/nathan/miniconda3/envs/jhub/lib/python3.6/site-packages/pandas/core/ops/array_ops.py\u001b[0m in \u001b[0;36mcomparison_op\u001b[0;34m(left, right, op)\u001b[0m\n\u001b[1;32m    245\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    246\u001b[0m     \u001b[0;32melif\u001b[0m \u001b[0mis_object_dtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlvalues\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 247\u001b[0;31m         \u001b[0mres_values\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcomp_method_OBJECT_ARRAY\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mop\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    248\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    249\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/nathan/miniconda3/envs/jhub/lib/python3.6/site-packages/pandas/core/ops/array_ops.py\u001b[0m in \u001b[0;36mcomp_method_OBJECT_ARRAY\u001b[0;34m(op, x, y)\u001b[0m\n\u001b[1;32m     55\u001b[0m         \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlibops\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvec_compare\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mravel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mop\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     56\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 57\u001b[0;31m         \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlibops\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mscalar_compare\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mravel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mop\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     58\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreshape\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     59\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/ops.pyx\u001b[0m in \u001b[0;36mpandas._libs.ops.scalar_compare\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: '>' not supported between instances of 'str' and 'int'"
     ]
    }
   ],
   "source": [
    "# For each cluster label, equivalent to iterating through each genome\n",
    "for cluster in range(len(clusterer.labels_)):\n",
    "    # check for genes of interest\n",
    "    row = vdf.iloc[cluster]\n",
    "    genes_of_interest = list(row[row > 0].index)\n",
    "    classification = association[cluster]\n",
    "    total_genes = list(set(gene_clusters.at[classification,clusterer.labels_[cluster]] + genes_of_interest))\n",
    "    gene_clusters.at[classification,clusterer.labels_[cluster]] = total_genes\n",
    "print(gene_clusters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    }
   ],
   "source": [
    "print(len(gene_clusters.iloc[0, 0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "At this point, our matrix contains genes that are not unique to 1 cluster and 1 association. For example, COG 2468 is found in both cluster 2 and 38 (vastly different clusters), while COG 1833 is found only in cluster 4, but is found in both plant association, non-plant associated, and soil-associated bacteria. Since we are only interested in the plant-associated bacteria, it makes sense that our genes of interest would not be found in non-plant or soil-associated equivalents. As well, our genes of interest should be unique to one cluster, as ideally the symbiotic relationship should not be found in other clusters. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Thus, we first remove the genes within a cluster that are found in multiple associations. We do this by looping through our matrix, and make a set of the differences between the plant associated genes and the soil and non-plant associated genes. This set comprises all the genes of plant-associated bacteria that would not be found in bacteria that are not plant associated. We repeat this to obtain genes unique to non-plant associated bacteria, as well as soil-associated bacteria.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "classifications = {0:\"PA\", 1:\"NPA\",  2:\"soil\"}\n",
    "\n",
    "# check each cluster\n",
    "for i in range(clusterer.labels_.max()+2):\n",
    "    # for each classification\n",
    "    for current_class in classifications.keys():\n",
    "        ind = list(range(3))\n",
    "        ind.pop(current_class)\n",
    "        # See if gene is unique to PA, NPA, Soil, within current cluster\n",
    "        unique = list(set(gene_clusters.iloc[current_class,i]) - set(gene_clusters.iloc[ind[0], i]) - set(gene_clusters.iloc[ind[1], i]))\n",
    "        gene_clusters.loc[classifications[current_class] + ' unique', i-1] = unique\n",
    "        #print(gene_clusters.at[j+3, clusterer.labels_[i]])\n",
    "print(gene_clusters.iloc[:, 1:])  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We save all the unique genes into a new dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "S_U = gene_clusters.iloc[3, 1:]\n",
    "PA_U = gene_clusters.iloc[4, 1:]\n",
    "NPA_U = gene_clusters.iloc[5, 1:]\n",
    "\n",
    "unique_genes = gene_clusters.iloc[3:, 1:]\n",
    "unique_genes.to_csv('unique_genes_association.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After transforming these sets into a Dataframe organized by association (index) and cluster number (column index), we iterate through all the clusters. For each cluster, we select genes unique to an association and compare those genes to all other genes in all other clusters. By taking their difference as a set, we ensure that our genes of interest are both unique to a cluster/taxonomy, and unique to an association."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# test all the PAU in one cluster and make sure they arent in another cluster\n",
    "copies_removed = S_U\n",
    "\n",
    "for cluster in range(len(copies_removed)):\n",
    "    all_genes_in_cluster = copies_removed.iloc[cluster]\n",
    "    original_number_of_genes = len(all_genes_in_cluster)\n",
    "    if (original_number_of_genes > 0):\n",
    "        inds = list(range(len(copies_removed)))\n",
    "        inds.pop(cluster)\n",
    "        # iteratively remove duplicates that are found in other clusters\n",
    "        for j in inds:\n",
    "            all_genes_in_cluster= list(set(all_genes_in_cluster) - set(copies_removed.iloc[j]))\n",
    "        #print(len(all_genes_in_cluster)/original_number_of_genes)\n",
    "        copies_removed.iat[cluster] = all_genes_in_cluster\n",
    "print(copies_removed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Comparison between new and old clusters, only for when .copy() is used since otherwise original df has been mutated\n",
    "for cluster in range(len(copies_removed)):\n",
    "    #print(len(copies_removed.iloc[cluster]), ' and ', len(S_U.iloc[cluster]))\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                            -1  \\\n",
      "PA           [COG0014, COG3913, COG0851, COG3898, COG1587, ...   \n",
      "NPA          [COG0014, COG3913, COG4481, COG0851, COG3898, ...   \n",
      "soil         [COG0014, COG3913, COG4481, COG0851, COG3898, ...   \n",
      "soil unique  [COG1887, COG3648, COG4940, COG2456, COG0615, ...   \n",
      "PA unique        [COG5482, COG4130, COG4189, COG1306, COG5659]   \n",
      "\n",
      "                                                             0  \\\n",
      "PA           [COG0014, COG4481, COG1587, COG1286, COG4838, ...   \n",
      "NPA          [COG0014, COG4481, COG1587, COG1286, COG4838, ...   \n",
      "soil         [COG0014, COG4481, COG1587, COG1286, COG4838, ...   \n",
      "soil unique  [COG4588, COG1212, COG4536, COG4115, COG3586, ...   \n",
      "PA unique    [COG2031, COG2170, COG3566, COG3012, COG3118, ...   \n",
      "\n",
      "                                                             1  \\\n",
      "PA           [COG0014, COG3913, COG0851, COG1587, COG1218, ...   \n",
      "NPA          [COG0014, COG3913, COG0851, COG1587, COG1218, ...   \n",
      "soil         [COG0014, COG3913, COG0851, COG1587, COG1218, ...   \n",
      "soil unique  [COG0410, COG1288, COG3123, COG5449, COG4233, ...   \n",
      "PA unique    [COG4521, COG5397, COG4259, COG3958, COG3575, ...   \n",
      "\n",
      "                                                             2  \\\n",
      "PA           [COG0014, COG1587, COG1218, COG1286, COG3808, ...   \n",
      "NPA          [COG0014, COG1587, COG1218, COG1286, COG3808, ...   \n",
      "soil         [COG0014, COG1587, COG1218, COG1286, COG3808, ...   \n",
      "soil unique  [COG3118, COG1169, COG3786, COG2717, COG1177, ...   \n",
      "PA unique    [COG5511, COG3453, COG5437, COG3837, COG3639, ...   \n",
      "\n",
      "                                                             3  \n",
      "PA           [COG0014, COG0851, COG3898, COG1587, COG1218, ...  \n",
      "NPA          [COG0014, COG3898, COG1587, COG1218, COG1286, ...  \n",
      "soil         [COG0014, COG3898, COG1587, COG1218, COG1286, ...  \n",
      "soil unique  [COG4315, COG5605, COG5586, COG4237, COG4608, ...  \n",
      "PA unique    [COG2933, COG4648, COG0851, COG3117, COG4255, ...  \n"
     ]
    }
   ],
   "source": [
    "for cluster in range(len(copies_removed)):\n",
    "     if len(copies_removed.iloc[cluster]) > 0:\n",
    "        #print('Cluster ', cluster, ' ', len(copies_removed.iloc[cluster]))\n",
    "        #print(copies_removed.iloc[cluster])\n",
    "        pass\n",
    "# Peek at unique gene cluster dataframe\n",
    "print(gene_clusters.head().iloc[:, 0:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                            0   \\\n",
      "soil unique  [COG4588, COG1212, COG4536, COG4115, COG3586, ...   \n",
      "PA unique    [COG2031, COG2170, COG3566, COG3012, COG3118, ...   \n",
      "NPA unique   [COG4483, COG4837, COG4477, COG3589, COG4851, ...   \n",
      "\n",
      "                                                            1   \\\n",
      "soil unique  [COG0410, COG1288, COG3123, COG5449, COG4233, ...   \n",
      "PA unique    [COG4521, COG5397, COG4259, COG3958, COG3575, ...   \n",
      "NPA unique   [COG3130, COG3703, COG2905, COG4764, COG5501, ...   \n",
      "\n",
      "                                                            2   \\\n",
      "soil unique  [COG3118, COG1169, COG3786, COG2717, COG1177, ...   \n",
      "PA unique    [COG5511, COG3453, COG5437, COG3837, COG3639, ...   \n",
      "NPA unique                                                  []   \n",
      "\n",
      "                                                            3   \\\n",
      "soil unique  [COG4315, COG5605, COG5586, COG4237, COG4608, ...   \n",
      "PA unique    [COG2933, COG4648, COG0851, COG3117, COG4255, ...   \n",
      "NPA unique   [COG4520, COG3154, COG1988, COG4233, COG5583, ...   \n",
      "\n",
      "                                                            4   \\\n",
      "soil unique  [COG0837, COG2933, COG3474, COG2839, COG0370, ...   \n",
      "PA unique    [COG3803, COG4758, COG1525, COG1416, COG4185, ...   \n",
      "NPA unique                [COG5528, COG0585, COG1344, COG2121]   \n",
      "\n",
      "                                                            5   \\\n",
      "soil unique  [COG5394, COG2229, COG1757, COG0813, COG1030, ...   \n",
      "PA unique    [COG3913, COG3844, COG0282, COG4227, COG1373, ...   \n",
      "NPA unique                                           [COG3940]   \n",
      "\n",
      "                                                            6   \\\n",
      "soil unique  [COG2855, COG0403, COG0819, COG1796, COG3562, ...   \n",
      "PA unique    [COG5664, COG3306, COG1535, COG4950, COG4102, ...   \n",
      "NPA unique   [COG1312, COG3845, COG1744, COG1079, COG5557, ...   \n",
      "\n",
      "                                                            7   \\\n",
      "soil unique      [COG1501, COG2440, COG0646, COG2111, COG1776]   \n",
      "PA unique    [COG4898, COG4731, COG5573, COG1988, COG5661, ...   \n",
      "NPA unique                                                  []   \n",
      "\n",
      "                                                            8   \\\n",
      "soil unique  [COG1213, COG4551, COG0625, COG0792, COG1626, ...   \n",
      "PA unique    [COG3474, COG1848, COG5519, COG4564, COG3491, ...   \n",
      "NPA unique                [COG5511, COG4619, COG3225, COG3736]   \n",
      "\n",
      "                                                            9   ...  \\\n",
      "soil unique  [COG0864, COG4677, COG0348, COG1882, COG4701, ...  ...   \n",
      "PA unique    [COG4328, COG1218, COG4908, COG3798, COG5605, ...  ...   \n",
      "NPA unique                                           [COG4519]  ...   \n",
      "\n",
      "                                                            31         32  \\\n",
      "soil unique  [COG4696, COG3360, COG2116, COG4859, COG4113, ...         []   \n",
      "PA unique    [COG3157, COG0741, COG0785, COG2205, COG1742, ...         []   \n",
      "NPA unique   [COG3118, COG1899, COG3579, COG4908, COG3487, ...  [COG0067]   \n",
      "\n",
      "                                                            33  \\\n",
      "soil unique                                                 []   \n",
      "PA unique                                                   []   \n",
      "NPA unique   [COG3147, COG0038, COG3265, COG3844, COG3214, ...   \n",
      "\n",
      "                                                            34  \\\n",
      "soil unique  [COG0658, COG3953, COG3283, COG3660, COG2152, ...   \n",
      "PA unique    [COG4696, COG4946, COG4677, COG3669, COG3898, ...   \n",
      "NPA unique   [COG4536, COG1427, COG3742, COG2869, COG4402, ...   \n",
      "\n",
      "                                                            35  \\\n",
      "soil unique                                                 []   \n",
      "PA unique    [COG1946, COG3214, COG3654, COG4413, COG3058, ...   \n",
      "NPA unique   [COG0068, COG2044, COG2143, COG5487, COG1633, ...   \n",
      "\n",
      "                                                            36  \\\n",
      "soil unique               [COG2409, COG3114, COG2853, COG3876]   \n",
      "PA unique    [COG5184, COG3271, COG3240, COG4804, COG1579, ...   \n",
      "NPA unique   [COG4657, COG4243, COG1437, COG4198, COG4659, ...   \n",
      "\n",
      "                                                            37  \\\n",
      "soil unique  [COG2402, COG5016, COG2138, COG1899, COG1388, ...   \n",
      "PA unique    [COG3136, COG1721, COG2144, COG4859, COG3915, ...   \n",
      "NPA unique   [COG4152, COG1528, COG1891, COG4328, COG1584, ...   \n",
      "\n",
      "                                                            38  \\\n",
      "soil unique  [COG4702, COG5552, COG1013, COG3080, COG3850, ...   \n",
      "PA unique    [COG3531, COG3826, COG1839, COG1615, COG4245, ...   \n",
      "NPA unique   [COG4129, COG3263, COG3016, COG1478, COG2311, ...   \n",
      "\n",
      "                                                            39  \\\n",
      "soil unique                                                 []   \n",
      "PA unique                                                   []   \n",
      "NPA unique   [COG0014, COG3913, COG2421, COG3015, COG1304, ...   \n",
      "\n",
      "                                                            40  \n",
      "soil unique  [COG3607, COG3411, COG3745, COG4964, COG1506, ...  \n",
      "PA unique                          [COG3591, COG3419, COG0447]  \n",
      "NPA unique   [COG4228, COG4143, COG5463, COG0791, COG4213, ...  \n",
      "\n",
      "[3 rows x 41 columns]\n"
     ]
    }
   ],
   "source": [
    "print(unique_genes)\n",
    "unique_genes.to_csv(\"UNQ2.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lastly, we transform our genes of interest into a Dataframe with 1 row for each gene, and summary statistics for its frequency."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compute the breakdown."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get breakdown of PA-NPA-soil ratio in each cluster, as well as cluster sizes\n",
    "breakdown= pd.DataFrame(data = np.zeros([3,clusterer.labels_.max()+1]), index =[\"NPA\", \"PA\", \"soil\"], \n",
    "                        columns = list(range(clusterer.labels_.max() + 1)))\n",
    "clusternums = list(range(clusterer.labels_.max()+1))\n",
    "cluster_sizes = {key: 0 for key in clusternums}\n",
    "\n",
    "print(breakdown)\n",
    "# for each genome\n",
    "for i in range(len(clusterer.labels_)):\n",
    "    # if the entry has been properly clustered\n",
    "    if clusterer.labels_[i] != -1:\n",
    "        assoc = 3\n",
    "        # assign the breakdown for the respective association, and the respective cluster number\n",
    "        breakdown.loc[hover_data.iloc[i, assoc], clusterer.labels_[i]] +=1\n",
    "        cluster_sizes[clusterer.labels_[i]] +=1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Parse metadata to replace COG IDs with gene descriptors."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_description = {}\n",
    "\n",
    "with open('cddid.tbl', 'r') as f:\n",
    "    fs = f.readlines()\n",
    "    for i in range(len(fs)):\n",
    "        line = fs[i].split(\"\\t\")\n",
    "        if 'COG' in line[1]:\n",
    "            gene_description[line[1]] = line[3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Initialize an empty dataframe and fill with summary statistics, computing the frequency by iterating through each cluster and classification."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pd' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-12fcad6fe490>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m table = pd.DataFrame(columns = [\"COG ID\", \"COG ID Description\", \"Cluster number\",\n\u001b[0m\u001b[1;32m      2\u001b[0m                                 \u001b[0;34m\"Frequency/# times found in cluster\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m                                 \"# PA in cluster\", \"# NPA\", \"# soil\", \"classification\", \"cluster size\"])\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;31m# Iterating through clusters 0-39\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'pd' is not defined"
     ]
    }
   ],
   "source": [
    "\n",
    "table = pd.DataFrame(columns = [\"COG ID\", \"COG ID Description\", \"Cluster number\",\n",
    "                                \"Frequency/# times found in cluster\", \n",
    "                                \"# PA in cluster\", \"# NPA\", \"# soil\", \"classification\", \"cluster size\"])\n",
    "\n",
    "# Iterating through clusters 0-39\n",
    "for cluster in range(clusterer.labels_.max()+1):\n",
    "    for classification in unique_genes.index:\n",
    "    # for each cluster-specific gene in the soil/pa/npa classification, append the class\n",
    "        for i in range(len(unique_genes.loc[classification, cluster])):\n",
    "            # get the key / descriptor for that gene + append to new table\n",
    "            gene = UNQ.loc[classification, cluster][i]\n",
    "            descr = 'N/A'\n",
    "            try:\n",
    "                descr = gene_description[gene]\n",
    "            except Exception as e:\n",
    "                pass\n",
    "            gene_num = 0\n",
    "            for l in range(len(vdf.columns)):\n",
    "                if vdf.columns[l] == gene:\n",
    "                    gene_num = l\n",
    "                    break\n",
    "            freq = sum([vdf.iloc[j, gene_num] for j in range(len(clusterer.labels_)) if clusterer.labels_[j] == cluster])\n",
    "            entry = [gene, descr, cluster, freq,breakdown.at[\"PA\",cluster],breakdown.at[\"NPA\",cluster],\n",
    "                     breakdown.at[\"soil\", cluster], classification, cluster_sizes[cluster]]\n",
    "            table.loc[gene] = entry\n",
    "print(table.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, let's save the table as a CSV."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "metadata": {},
   "outputs": [],
   "source": [
    "table.to_csv(\"singular_genes_by_association.csv\") "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "metagenomics samples that are revelant to the PA genes (the bins/reassembled genome that contain the genes)\n",
    "1. find microbes that are PA\n",
    "2. break down by metagenomics-do the binning, and then\n",
    "3. look for distribution/enrichment of PA genes\n",
    "\n",
    "\n",
    "Grow bacteria , extract DNA, \n",
    "Symbiotic rels of bacteria in dirt, \n",
    "Cant isolate the bugs in soil so just take all the DNA you can find, \n",
    "Bin together similar taxons, see what comes out, assembly and maximizes coverage\n",
    "\n",
    "Are those PA genes being found in our genes? Using metagenomics as a type of validation\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
